# Operators

## Definitions

### Operators & Derived Functions

We have already seen some operators: _reduce_ (described in [here](./Some-Primitive-Functions.ipynb#Reduce)), _axis_ (described in [here](./Some-Primitive-Functions.ipynb#Axis-Specification)), and _each_ (described in [here](./Nested-Arrays-Continued.ipynb#Each).
Let us define precisely what they are:

 - there are built-in (_primitive_) operators and user-defined operators;
 - an _operator_ is similar to a function, but rather than working on arrays to produce a result which is also an array, an operator works on functions (and sometimes, arrays) to produce a new function;
 - the new function generated by the operator and its argument(s) is called a _derived function_. The derived function can be applied to arrays in the same way as any other function;
 - the arguments passed to the operator are often referred to as _operands_, to distinguish them from the arguments to the derived function;
 - monadic operators take a single operand on their **left**. This is in contrast to monadic functions, which take their argument on the right;
 - dyadic operators have two operands, one on each side. The operands to an operator are usually functions, but it is not uncommon for user-defined operators to take on function and one array operand;
 - the derived function, in turn, can be monadic, dyadic, or ambivalent; and
 - neither of the functions supplied as arguments to an operator, nor the resultant derived function, can be niladic.

For example, in the expression below, the operator _reduce_ (`/`) _operates on_ the function _plus_ to produce the derived function _plus reduce_.
This derived function is then applied to `3 5 6` to produce the result `14`:

In [1]:
+/3 5 6

14

<!-- begin beware style=warning -->
***Beware***:

 > You must not be confused by the fact that some symbols are used to represent both a function and an operator, such as `/` and `\`, for example.
<!-- end -->

Let us compare two expressions:

 1. using `/` as the function _compress_, which takes two array arguments:

In [2]:
1 1 0 1 0 / 6 2 9 4 5

6 2 4

 2. using `/` as the operator _reduce_, which takes a function as a left operand:

In [3]:
+/ 6 2 9 4 5

26

The association of `+` with `/` creates a _derived function_ which could be parenthesised as `(+/)`, even though it is not necessary to do so.

For clarification, we can define a synonym for the derived function:

In [4]:
Sum ← +/

Until now, we have only considered `+/` (or `Sum`) as a monadic derived function:

In [5]:
Sum 6 2 9 4 5

26

But we shall soon see that it may also be used as a dyadic function:

In [6]:
2 Sum 6 2 9 4 5

8 11 13 9

So, we can say that this _derived function_ is _ambivalent_.

### Sequences of Operators

_Derived functions_ behave exactly like plain primitive functions.
So, they can be the argument of a second (and a third, ...) operator:

In [7]:
+/¨ (3 4 6)(4 9 7 1)(3 1)

13 21 4

The left argument of the operator _each_ `¨` is the derived function `+/`, so we could have written:

In [8]:
Sum¨ (3 4 6)(4 9 7 1)(3 1)

13 21 4

Now, suppose that we no longer want to add up vectors, but three small matrices instead:

In [9]:
⎕← A ← 2 3⍴⍳6

1 2 3
4 5 6

In [10]:
⎕← B ← 4 2⍴1 0 0 1 0 1 1 0

1 0
0 1
0 1
1 0

In [11]:
⎕← C ← 3 5⍴8 3 4 2 0 0 3 5 1 7 3 6 2 1 7

8 3 4 2 0
0 3 5 1 7
3 6 2 1 7

Because they are matrices, we must specify the axis along which we add them up.
Of course, we could use the two symbols `/` and `⌿`, but if the arrays had been of a higher rank, an explicit axis specification might have been necessary.
It could also be that we just prefer to use the explicit axis specification.
If so, a third level of operator⁽*⁾ can be added:

<!-- begin footnote -->
***Note**:

 > Although the _axis specification_ shares some properties with operators, it is a special syntactical element and not really an operator. See [below](#Axis) for more information.
<!-- end -->

In [12]:
+/[2]¨A B C

┌────┬───────┬────────┐
│6 15│1 1 1 1│17 16 19│
└────┴───────┴────────┘

In [13]:
+/[1]¨A B C

┌─────┬───┬─────────────┐
│5 7 9│2 2│11 12 11 4 14│
└─────┴───┴─────────────┘

When in doubt regarding what is the function operand to which operator, try parenthesising everything successively, to make it clearer what derived functions come from where.

In the two examples above, the first operator is `/`, then the second "operator" is `[]`, and the third operator is `¨`:

In [14]:
(((+/)[2])¨)A B C

┌────┬───────┬────────┐
│6 15│1 1 1 1│17 16 19│
└────┴───────┴────────┘

### List of Built-in Operators

Dyalog APL has a rich set of built-in operators.
You will find a full list with detailed syntax and examples in [the Appendix](./Appendices.ipynb#Dyalog-APL-Operators).

## More About Some Operators You Already Know

### Reduce

Up to now, we have used the operator _reduce_ with rather basic functions (`+`, `×`, `⌈`, `∧`), but it can also be used, less obviously, with functions like _reshape_, _compress_, and _replicate_.
In these cases, the derived function typically takes a 2-item nested vector as its argument, and the effect is to insert the function (the operand to the operator) between the two items of this vector.

Just remember that, since `+/ (2 4 3)(7 1 5)` is equivalent to `⊂(2 4 3) + (7 1 5)`, then `⍴/ (2 4 3)(7 1 5)` is also equivalent to `⊂(2 4 3) ⍴ (7 1 5)`.

Here is an example of a _reduction by reshape_:

In [15]:
⍴/(2 5)(3 1 9 4 1 0 7)

┌─────────┐
│3 1 9 4 1│
│0 7 3 1 9│
└─────────┘

This _looks_ very similar to

In [16]:
2 5⍴3 1 9 4 1 0 7

3 1 9 4 1
0 7 3 1 9

But we can see the results are different.
The result of the _reduction by reshape_ is not a matrix, but a scalar containing a _nested_ matrix, for the reason already seen in [a previous section](./Nested-Arrays-Continued.ipynb#Reduction): the reduction of a vector always gives a scalar.

Now, here is a _reduction by compression_, another by _replication_, and one by _index of_:

In [17]:
//(1 1 0 1 0 1 1)'Strange'

┌─────┐
│Stage│
└─────┘

In [18]:
//(1 1 0 4 0 1 2)'Strange'

┌─────────┐
│Staaaagee│
└─────────┘

In [19]:
⍳/(2 6 1 7)(2 4⍴3 7 8 4 2 5 6 0)

┌───────┐
│5 4 5 5│
│1 5 2 5│
└───────┘

### n-Wise Reduce

#### Elementary Definition

The derived functions of _reduce_ can be used with two arguments.
When the second argument is present, the form is called _n-wise reduce_.

When applied to vectors, _n-wise reduce_ has the syntax `r ← n F/ vector`, where `F` denotes a dyadic function.

This special kind of _reduce_ splits the vector into overlapping slices of length equal to `n`, reduces each slice using the specified function `F`, and then catenates the results together.

So, for example, `2 ×/ 8 10 7 2 6 11` starts by creating overlapping slices of length `2`:

In [20]:
(8 10)(10 7)(7 2)(2 6)(6 11)

┌────┬────┬───┬───┬────┐
│8 10│10 7│7 2│2 6│6 11│
└────┴────┴───┴───┴────┘

Then, we apply the reduction `×/` on each slice and catenate everything:

In [21]:
(×/8 10),(×/10 7),(×/7 2),(×/2 6),(×/6 11)

80 70 14 12 66

You can verify that this is, indeed, the result we get:

In [22]:
2 ×/ 8 10 7 2 6 11

80 70 14 12 66

As another example, we can explain the result of

In [23]:
3 +/ 8 10 7 2 6 11

25 19 15 19

because we create overlapping slices of length `3`, apply _plus-reduce_ on each slice, and then catenate everything together:

In [24]:
(+/8 10 7),(+/10 7 2),(+/7 2 6),(+/2 6 11)

25 19 15 19

The length of the result vector is `(1+≢vector)-n`.

In the two examples above, we did not really need to do the catenation explicitly, because the results of applying `×/` and `+/` on each slice were simple scalars.
However, if we try an example where the reduction gives a nested result, we will see that we _do_ need to catenate everything together.

Here is an example of an _n-wise catenate reduction_:

In [25]:
2 ,/ 8 10 7 2 6 11

┌────┬────┬───┬───┬────┐
│8 10│10 7│7 2│2 6│6 11│
└────┴────┴───┴───┴────┘

If we create the overlapping slices by hand and do not catenate the results together, we get a result that is nested too deep:

In [26]:
(,/8 10)(,/10 7)(,/7 2)(,/2 6)(,/6 11)

┌──────┬──────┬─────┬─────┬──────┐
│┌────┐│┌────┐│┌───┐│┌───┐│┌────┐│
││8 10│││10 7│││7 2│││2 6│││6 11││
│└────┘│└────┘│└───┘│└───┘│└────┘│
└──────┴──────┴─────┴─────┴──────┘

Thus, we do have to catenate the results after applying the reduction to each slice:

In [27]:
(,/8 10),(,/10 7),(,/7 2),(,/2 6),(,/6 11)

┌────┬────┬───┬───┬────┐
│8 10│10 7│7 2│2 6│6 11│
└────┴────┴───┴───┴────┘

#### Full Definition

The general syntax is `r ← n F/[axis] array`, where `F` stands for any dyadic function.

 - the `array` is split into slices along the specified `axis`;
 - the left argument `n` can be positive (as in the examples above), zero, or negative;
 - if `n` is positive, _reduce_ is applied to slices of length equal to `n`;
 - if `n` is zero, the result is an array with the same shape as _array_, except that its length along the axis selected by `axis` is incremented by 1 and filled with the _identity item_ for the function `F`. This is explained in [a Specialist Section](#application-to-n-wise-reduction); and
 - if `n` is negative, each slice is reversed before _reduce_ is applied.

Here are some examples which use the matrix below:

In [28]:
tam ← 3 5⍴2 3 5 8 8 4 6 2 5 9 1 4 9 7 8

Find the largest items of 2 adjacent columns:

In [29]:
2 ⌈/ tam

3 5 8 8
6 6 5 9
4 9 9 8

Add up pairs of adjacent rows:

In [30]:
2 +⌿ tam

6 9 7 13 17
5 10 11 12 17

Return a matrix with one more column, filled with zeroes (identity item of addition):

In [31]:
0 +/ tam

0 0 0 0 0 0
0 0 0 0 0 0
0 0 0 0 0 0

Return a matrix with one more row, filled with ones (identity item of multiplication):

In [32]:
0 ×/[1] tam

1 1 1 1 1
1 1 1 1 1
1 1 1 1 1
1 1 1 1 1

Obtain the differences between adjacent values `(14-11)(15-14)...`:

In [33]:
¯2 -/ 11 14 15 21 23 30 28 34

3 1 6 2 7 ¯2 6

### Axis

Strictly speaking, _axis_ is not an operator.
It has different syntax (consisting of two brackets enclosing a numeric value to the right of a function) and applies in different ways depending on the function that it modifies (its operand).
However, applying a function "with _axis_" does apply a transformation and produces a derived function, and it is common to think of _axis_ as an operator.

It is possible to use _axis_ with any of the scalar dyadic functions.
This can be useful, for example, to add the items of a vector to each of the rows of a matrix, or multiply the columns of a matrix by different values:

In [34]:
tam

2 3 5 8 8
4 6 2 5 9
1 4 9 7 8

In [35]:
tam +[1] 8 6 9

10 11 13 16 16
10 12 8 11 15
10 13 18 16 17

In [36]:
tam ×[2] 2 5 0 2 1

4 15 0 16 8
8 30 0 10 9
2 20 0 14 8

The list of all scalar dyadic functions is given in [an appendix](./Appendices.ipynb#Scalar-Functions).

The following functions can also use _axis_:

| Monadic Function | Description |
| :- | :- |
| `↑` and `↓` | _mix_ and _split_ |
| `⌽` `⊖` | _reverse_ |
| `,` | _ravel with axis_ |
| `⊂` | _enclose with axis_, _partitioned enclose_ |
| `⊂` and `⊃` | if `⎕ML>1`, APL2-like _split_ and _mix_, c.f. [this section](./Nested-Arrays-Continued.ipynb#Compatibility-and-Migration-Level) |

| Dyadic Function | Description |
| :- | :- |
| `+` `×` `⌈` `∧` `≤` etc... | all scalar dyadic functions |
| `↑` and `↓` | _take_ and _drop_ |
| `/` and `⌿` | _compress_ and _replicate_ |
| `\` and `⍀` | _expand_ and _scan_ (see next section) |
| `⌽` `⊖` | _rotate_ |
| `,` `⍪` | _catenate_ |
| `,` `⍪` | _laminate_ |
| `⊂` | _partitioned enclose_ |

## Scan

### Definition

_Scan_ is represented by the symbol `\` or `⍀`.
Its most general syntax is `r ← F\[axis] array`, where `F` stands for any appropriate dyadic function.

To understand how it works, let us apply it to a vector.

The nth item of `F\vector` is equal to `F/n↑vector`. A worked example follows:

In [37]:
+\ 3 6 1 8 5

3 9 10 18 23

 1. the 1st item is equal to `+/3`, which is `3`;
 2. the 2nd item is equal to `+/3 6`, which is `9`;
 3. the 3rd item is equal to `+/3 6 1`, which is `10`;
 4. the 4th item is equal to `+/3 6 1 8`, which is `18`; and
 5. the 5th item is equal to `+/3 6 1 8 5`, which is `23`.

Try to use a reasoning similar to the one above to understand the result of the _times scan_ shown below:

In [38]:
×\ 3 6 1 8 5

3 18 18 144 720

Warning!
It would be a mistake to always try to deduce the value of each item in the result from its immediate left neighbour.
While it is possible to do this for commutative functions like addition and multiplication, it is not appropriate for non-commutative functions like subtraction.

In fact, the result of `-\ 3 6 1 8 5` is **not** `3 ¯3 ¯4 ¯12 ¯17`, but something else entirely:

In [39]:
-\ 3 6 1 8 5

3 ¯3 ¯2 ¯10 ¯5

 1. the 1st item is equal to `-/3`, which is `3`;
 2. the 2nd item is equal to `-/3 6`, which is `¯3`;
 3. the 3rd item is equal to `-/3 6 1`, which is `¯2`, the result of `3-6-1` and **not** `(3-6)-1`;
 4. the 4th item is equal to `-/3 6 1 8`, which is `¯10`; and
 5. the 5th item is equal to `-/3 6 1 8 5`, which is `¯5`.

So, be careful when using _scan_ with non-commutative functions.

When applied to matrices or higher-rank arrays, _scan_ works along the specified axis.
If the axis specification is omitted, `\` works along the _last_ axis and `⍀` works along the _first_ axis.

In [40]:
+\[2]tam  ⍝ same as +\tam

2 5 10 18 26
4 10 12 17 26
1 5 14 21 29

In [41]:
+\[1]tam  ⍝ same as +⍀tam

2 3 5 8 8
6 9 7 13 17
7 13 16 20 25

### Scan with Binary Values

_Scan_ is very useful when applied to binary values.

In [42]:
∨\ 0 0 0 0 1 1 0 1 0 0 1 1

0 0 0 0 1 1 1 1 1 1 1 1

Because the function _or_ gives the result `1` as soon as one of its arguments is `1`, _or-scan_ repeats the first `1` up to the end of the vector.
In a way, you can see `∨\` as a knife spreading butter from left to right, and the `1` is the butter.

Other interesting patterns can be obtained by changing the function used.
For example, you can get the effect of "a knife spreading butter", where the `0` is the butter, if you use `∧\` instead of `∨\`:

In [43]:
∧\ 1 1 1 1 0 1 1 0 0 1 1 0

1 1 1 1 0 0 0 0 0 0 0 0

In the example above, as soon as _and-scan_ finds a `0`, everything else turns into a `0`.

The _less-than-scan_ marks the position of the first `1` and the _less-than-or-equal-scan_ marks the position of the first `0`:

In [44]:
<\ 0 0 0 0 1 1 0 1 0 0 1 1

0 0 0 0 1 0 0 0 0 0 0 0

In [45]:
≤\ 1 1 1 1 0 1 1 0 0 1 1 0

1 1 1 1 0 1 1 1 1 1 1 1

### Applications

_Scan_ can be used to solve common problems in a very simple way:

#### Inflate Values

Someone forecasts investments in a foreign country for the next five years:

In [46]:
inv ← 2000 5000 6000 4000 2000

But the country in question suffers from inflation, and the inflation rates are forecasted as follows:

In [47]:
inf ← 2.6 2.9 3.4 3.1 2.7

The cumulative sequence of these inflation rates can be calculated by multiplying them all with a _multiply-scan_:

In [48]:
7 3⍕ ×\ 1+inf÷100

1.026 1.056 1.092 1.125 1.156

Now, the investments expressed in "future values" would be:

In [49]:
9 2⍕ inv × ×\1+inf÷100

2052.00 5278.77 6549.90 4501.96 2311.76

Finally, the year after year cumulative investment may be obtained by a _plus-scan_:

In [50]:
9 2⍕ +\ inv × ×\1+inf÷100

2052.00 7330.77 13880.67 18382.63 20694.39

As you can see, we employed two _scans_ in the same expression.

#### Remove Leading/Trailing Blanks

One often has to remove leading (or trailing) blanks from a character vector.
We can use the _or-scan_ to do it.
The details of the method are shown here:

In [51]:
lb ← '    Remove my 4 leading blanks.'
lb≠' '

0 0 0 0 1 1 1 1 1 1 0 1 1 0 1 0 1 1 1 1 1 1 1 0 1 1 1 1 1 1 1

In [52]:
∨\ lb≠' '

0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1

In [53]:
(∨\ lb≠' ')/lb

Remove my 4 leading blanks.

This can be coded into a small utility function:

In [54]:
CutBlanks ← {(∨\' '≠⍵)/⍵}

This expression is recognised by Dyalog APL as an _idiom_ and is thus very fast.

To remove trailing blanks, it would suffice to reverse the vector, remove leading blanks as above, and then reverse it back again.

## Outer Product

### Definition

Imagine that you have calculated the multiplication table for the integers 1 to 9; you could present it like this:

$$\begin{array}{|c|rrrrrrrrr|}
\hline
\color{red}\times & 1 & 2 & 3 & 4 & 5 & 6 & \color{red} 7 & 8 & 9 \\
\hline
1 & 1 & 2 & 3 & 4 & 5 & 6 & 7 & 8 & 9 \\
2 & 2 & 4 & 6 & 8 & 10 & 12 & 14 & 16 & 18 \\
\color{red} 3 & 3 & 6 & 9 & 12 & 15 & 18 & \color{red} {21} & 24 & 27 \\
4 & 4 & 8 & 12 & 16 & 20 & 24 & 28 & 32 & 36 \\
\text{etc.} & \text{etc.} & & & & & & & & & & \\
9 & 9 & 18 & 27 & 36 & 45 & 54 & 63 & 72 & 81 \\
\hline
\end{array}$$

The task of calculating this table consists of taking pairs of items of two vectors (the column and row headings) and combining them with the function at the top left.
For example, `3` times `7` gives `21` (highlighted in red above).
Once the operation has been repeated for all the possible pairs, one obtains what is called, in APL, the _outer product_.

We can change the values and replace the multiplications with additions:

$$\begin{array}{|c|rrrrrr|}
\hline
\color{red}+ & 8 & 5 & 15 & \color{red} 9 & 11 & 40\\
\hline
5 & 13 & 10 & 20 & 14 & 16 & 45 \\
\color{red} 4 & 12 & 9 & 19 & \color{red} 13 & 15 & 44 \\
10 & 18 & 15 & 25 & 19 & 22 & 50 \\
3 & 11 & 8 & 18 & 12 & 14 & 43 \\
\hline
\end{array}$$

The _outer product_ operator looks like `∘.F`, where `F` is an appropriate dyadic function.
The symbol `∘` is a _jot_ and you can type it with <kbd>APL</kbd>+<kbd>j</kbd>.

For example, for the multiplication table you can use `∘.×`:

In [55]:
(⍳9) ∘.× ⍳9

1 2 3 4 5 6 7 8 9
2 4 6 8 10 12 14 16 18
3 6 9 12 15 18 21 24 27
4 8 12 16 20 24 28 32 36
5 10 15 20 25 30 35 40 45
6 12 18 24 30 36 42 48 54
7 14 21 28 35 42 49 56 63
8 16 24 32 40 48 56 64 72
9 18 27 36 45 54 63 72 81

And for the addition table above, you can use `∘.+`:

In [56]:
5 4 10 3 ∘.+ 8 5 15 9 11 40

13 10 20 14 16 45
12 9 19 13 15 44
18 15 25 19 21 50
11 8 18 12 14 43

As you can see, the _outer product_ is also slightly special, in that the operand function `F` goes on the right of `∘.`, and not on the left.

Also, notice that the left column of the table is the left argument vector to the function derived from the _outer product_ and the top row is the right argument vector.
In fact, in the expression `r ← left ∘.F right`, the shape of the result `r` is `(⍴r) ≡ (⍴left),⍴right`.

### Extensions

#### Other Functions

The function used in an _outer product_ can be any primitive or user-defined dyadic function, so _outer product_ is an operator of amazing power.

Imagine you have written a little function to calculate the length of the hypotenuse of a right-angled triangle from the lengths of the other 2 sides given as the left and right argument:

In [57]:
Hypo ← {((⍺*2)+(⍵*2))*0.5}
3 Hypo 4

5

You can test it on a number of combinations of lengths in one expression like this:

In [58]:
8 3⍕ 3 6 12 ∘.Hypo 4 1 8 7 5

5.000 3.162 8.544 7.616 5.831
 7.211 6.083 10.000 9.220 7.810
 12.649 12.042 14.422 13.892 13.000

Now, let us have some fun with relational functions:

In [59]:
(⍳5) ∘.= (⍳5)

1 0 0 0 0
0 1 0 0 0
0 0 1 0 0
0 0 0 1 0
0 0 0 0 1

In [60]:
(⍳5) ∘.< (⍳5)

0 1 1 1 1
0 0 1 1 1
0 0 0 1 1
0 0 0 0 1
0 0 0 0 0

In [61]:
(⍳5) ∘.≥ (⍳5)

1 0 0 0 0
1 1 0 0 0
1 1 1 0 0
1 1 1 1 0
1 1 1 1 1

We shall study some applications of _outer product_ like `∘.<` or `∘.⌊` in [a later subsection](#Applications-of-Outer-Product).

Whenever the function `F` does not produce a scalar, the _outer product_ `∘.F` produces a nested array. This is the case with _outer products_ like `∘.⍴`, `∘.,`, or `∘./`:

In [62]:
3 4 2 ∘.⍴ 6 3 7

┌───────┬───────┬───────┐
│6 6 6 │3 3 3 │7 7 7 │
├───────┼───────┼───────┤
│6 6 6 6│3 3 3 3│7 7 7 7│
├───────┼───────┼───────┤
│6 6 │3 3 │7 7 │
└───────┴───────┴───────┘

In [63]:
3 0 2 ∘./ 5 1 7

┌─────┬─────┬─────┐
│5 5 5│1 1 1│7 7 7│
├─────┼─────┼─────┤
│ │ │ │
├─────┼─────┼─────┤
│5 5 │1 1 │7 7 │
└─────┴─────┴─────┘

In [64]:
3 1 2 ∘., 6 3 0 7

┌───┬───┬───┬───┐
│3 6│3 3│3 0│3 7│
├───┼───┼───┼───┤
│1 6│1 3│1 0│1 7│
├───┼───┼───┼───┤
│2 6│2 3│2 0│2 7│
└───┴───┴───┴───┘

In [65]:
3 2 4 ∘.↑ 5 8 4

┌───────┬───────┬───────┐
│5 0 0 │8 0 0 │4 0 0 │
├───────┼───────┼───────┤
│5 0 │8 0 │4 0 │
├───────┼───────┼───────┤
│5 0 0 0│8 0 0 0│4 0 0 0│
└───────┴───────┴───────┘

#### Other Shapes and Types of Data

So far, we have applied _outer product_ to numeric vectors; it can, of course, also be used with character data and higher-rank arrays.
When applied to higher rank arrays, the result becomes very big quickly, because each item of the left array has to be combined with each item of the right one.

Remember, in the expression `r ← left ∘.F right`, the shape of `r` is equal to `(⍴left),(⍴right)`.

In [66]:
⎕← left ← ↑'DIMITRI' 'GUNTHER'

DIMITRI
GUNTHER

In [67]:
right ← 'VERONICA'

Now, we wish to study the result of `left ∘.= right`.
To help you visualise the comparisons being made, we catenate the left argument matrix on the left of the result and catenate the right argument vector on top:

In [68]:
(2 9⍴' ',right)⍪[2]left,left ∘.= right

V E R O N I C A
D 0 0 0 0 0 0 0 0
I 0 0 0 0 0 1 0 0
M 0 0 0 0 0 0 0 0
I 0 0 0 0 0 1 0 0
T 0 0 0 0 0 0 0 0
R 0 0 1 0 0 0 0 0
I 0 0 0 0 0 1 0 0
 
 V E R O N I C A
G 0 0 0 0 0 0 0 0
U 0 0 0 0 0 0 0 0
N 0 0 0 0 1 0 0 0
T 0 0 0 0 0 0 0 0
H 0 0 0 0 0 0 0 0
E 0 1 0 0 0 0 0 0
R 0 0 1 0 0 0 0 0

The left argument is a matrix with shape `2 7` and the right argument is a vector with shape `8`, so the result is a 3D array with shape `2 7 8`.
Each of the two major cells corresponds to comparing one of the names in the left argument to the name `'VERONICA'`.

### Applications of Outer Product

#### Exhaustive Search

Because _outer product_ uses a dyadic function to combine **all** items of the left argument with **all** items of the right argument, _outer product_ is often used when some kind of exhaustive computation needs to be done.
One such example is that of exhaustive search.

As an example, suppose you want to figure out if there is a way to add one number from the vector `5 1 16 42 63 7 10` to another number from the vector `24 45 18 31 29 43 67` to get `73`.

With _outer product_, this is quite an easy question to answer:

In [69]:
5 1 16 42 63 7 10∘.+24 45 18 31 29 43 67

29 50 23 36 34 48 72
25 46 19 32 30 44 68
40 61 34 47 45 59 83
66 87 60 73 71 85 109
87 108 81 94 92 106 130
31 52 25 38 36 50 74
34 55 28 41 39 53 77

In [70]:
73∊5 1 16 42 63 7 10∘.+24 45 18 31 29 43 67

1

If you flip the arguments to _membership_ and use _where_, you can find the position(s) where `73` is:

In [71]:
⍸(5 1 16 42 63 7 10∘.+24 45 18 31 29 43 67)∊73

┌───┐
│4 4│
└───┘

This pattern of exhaustive computations is fairly common, and although it generally is not the most computationally efficient way of solving a problem, it is generally fast enough to prototype as a first approach.

#### Draw a Bar Chart

Imagine that you have to represent a list of values with a bar chart.
Perhaps you will use dedicated graphical software, and you'd be right, but just have a look at this elegant solution, which again uses an _outer product_.

Here is the list of values that we want to chart:

In [72]:
nums ← 1 3 0 7 9 8 5 4 2 3 1

Let us first calculate the vertical scale.
It is made of the integers from 9 to 1 in reverse order and can be obtained by:

In [73]:
⌽⍳⌈/nums

9 8 7 6 5 4 3 2 1

Then, let us compare this scale to the values; an _outer product_ will build columns of `1`s up to the correct height:

In [74]:
(⌽⍳⌈/nums) ∘.≤ nums

0 0 0 0 1 0 0 0 0 0 0
0 0 0 0 1 1 0 0 0 0 0
0 0 0 1 1 1 0 0 0 0 0
0 0 0 1 1 1 0 0 0 0 0
0 0 0 1 1 1 1 0 0 0 0
0 0 0 1 1 1 1 1 0 0 0
0 1 0 1 1 1 1 1 0 1 0
0 1 0 1 1 1 1 1 1 1 0
1 1 0 1 1 1 1 1 1 1 1

Finally, to draw the graph, we can index a two-character vector, exactly as we did in [a previous section](./Data-and-Variables.ipynb#The-Shape-of-the-Result):

In [75]:
' ⎕'[1+(⌽⍳⌈/nums)∘.≤nums]

⎕ 
 ⎕⎕ 
 ⎕⎕⎕ 
 ⎕⎕⎕ 
 ⎕⎕⎕⎕ 
 ⎕⎕⎕⎕⎕ 
 ⎕ ⎕⎕⎕⎕⎕ ⎕ 
 ⎕ ⎕⎕⎕⎕⎕⎕⎕ 
⎕⎕ ⎕⎕⎕⎕⎕⎕⎕⎕

#### Decreasing Refunding

Some students have spent money to buy expensive books for their studies:

In [76]:
exp ← 740 310 1240 620 800 460 1060

Their university agrees to refund them, but places the following limits on the refunding rates:

| Expense range | Refund rate |
| :- | :- |
| 0 - 500 | 80% |
| 500 - 900 | 50% |
| 900+ | 0% |

We could say exactly the same thing in a somewhat different way:

Expenses from 0 to 900 have a refund rate of 50% and expenses up to 500 get an **additional** 30%.

Even if this rule may seem strange, both methods give the same result.
For example, a student who spent 740€ would get:

 - using the initial table, 80% of 500 plus 50% of 240:

In [77]:
(0.8×500)+0.5×240

520

 - using the reworded rule, 50% of 740 plus 30% of 500:

In [78]:
(0.5×740)+0.3×500

520

Now, let us limit the expenses to the given maxima:

In [79]:
exp ∘.⌊ 900 500

740 500
310 310
900 500
620 500
800 500
460 460
900 500

The first column of the result above contains the expenses capped at 900 (which get a refund of 50%) and the second column contains the expenses capped at 500 (which get an additional 30%).

So, according to our modified rule, we must pay 50% of the first column plus 30% of the second, which we can do by multiplying the columns with `0.5 0.3` (using an _axis_ operator) and then adding them:

In [80]:
+/ (exp∘.⌊900 500) ×[2] 0.5 0.3

520 248 600 460 550 368 600

And the total refund is, of course:

In [81]:
+/ +/(exp∘.⌊900 500)×[2]0.5 0.3

3346

If we laminate the original vector, we can see the expenses and the refunding:

In [82]:
exp,[.5] +/(exp∘.⌊900 500)×[2]0.5 0.3

740 310 1240 620 800 460 1060
520 248 600 460 550 368 600

!["APLer applying sunscreen outside."](../res/Outer_Product.png)

### Outer Product Exercise

<!-- exercise ex-complex-refunding -->
***Exercise 1***:

Let us try to generalise the method used above to compute refunds.

In our example, we had chosen a very simple case, because we had only two slices, and all students used the same scale.
Let us now imagine a slightly more complex case:

 - the students are classified in three categories, which have different refunding rates; and
 - we now have four different expense ranges.

The new conditions are expressed with the traditional notation in the table below:

| Category \ Range | 0 to 600 | 600 to 1.100 | 1.100 to 1.500 | 1.500 to 2.000 |
| :- | :- | :- | :- | :- |
| Category 1 | 100% | 100% | 80% | 50% |
| Category 2 | 100% |  70% | 30% | 10% |
| Category 3 |  80% |  60% | 20% |  5% |

```APL
limits ← 600 1100 1500 2000
rates ← 3 4⍴100 100 80 50 100 70 30 10 80 60 20 5
```

Define a function `Refund` to solve this problem.
The function should take the limits vector, the rates matrix, the expenses vector, and the categories vector, as right arguments.
The return value should be a vector with the refunded amount to each student.

Using loops is strictly prohibited and may be punished with high severity!

Make use of the variables `expenses` and `categories` defined below and verify your solution by comparing it to the values shown below.
<!-- end -->

In [83]:
⎕RL ← 73
expenses ← ?350⍴2500
categories ← ?350⍴3
⎕← 2 10↑categories,[.5]expenses

1 2 2 3 2 3 2 3 1 2
2121 1936 682 818 1922 1419 1389 465 1829 221

In [84]:
10↑Refund limits rates categories expenses

VALUE ERROR: Undefined name: rates
      10↑Refund limits rates categories expenses
                       ∧


In [85]:
+/Refund limits rates categories expenses

VALUE ERROR: Undefined name: rates
      +/Refund limits rates categories expenses
                      ∧


## Inner Product

_Inner product_ is a generalisation of what mathematicians call _matrix product_, a tool considered by most students as extremely abstract, full of bizarre notations, like $\sum a_{ij}b_{jk}$, and obviously far removed from everyday problems.
You will discover that:

 - the concept is really simple, nearly obvious; and
 - it can be applied to many real life problems.

A simple example will help us.

### A Concrete Situation

A company intends to open a series of hotels and resorts in four countries.
This requires serious investments over a period of five years.
The following table shows these investments (in millions of dollars, of course!):

| Country vs Year | Year 1 | Year 2 | Year 3 | Year 4 | Year 5|
| :- | -: | -: | -: | -: | -: |
| Greece | 120 | 100 | 40 | 20 | 0 |
| Brazil | 200 | 150 | 100 | 120 | 200 |
| Egypt | 50 | 120 | 220 | 350 | 600 |
| Argentina | 0 | 80 | 100 | 110 | 120 |

These figures are contained in a matrix called `invest`:

In [86]:
invest ← 120 100 40 20 0 200 150 100 120 200 50 120
⎕← invest ← 4 5⍴invest,220 350 600 0 80 100 110 120

120 100 40 20 0
200 150 100 120 200
 50 120 220 350 600
 0 80 100 110 120

These investments will be supported by the company itself plus two banks, each taking a certain percentage of the total, depending on the evaluation of each project.
The following table shows how the risks are shared:

| Entity vs Country | Greece | Brazil | Egypt | Argentina |
| :- | -: | -: | -: | -: |
| Bank 1 | 50 | 10 | 20 | 30 |
| Bank 2 | 20 | 60 | 40 | 30 |
| Company | 30 | 30 | 40 | 40 |

Those percentages are contained in a matrix named `percent`:

In [87]:
⎕← percent ← 3 4⍴50 10 20 30 20 60 40 30 30 30 40 40

50 10 20 30
20 60 40 30
30 30 40 40

We would like to calculate, year by year, how much each of the 3 partners is engaged in this project.
For example, let us try to evaluate the contribution of Bank 2 during Year 3:

| Country | Project valuation | Stake | Total invested |
| :- | -: | -: | -: |
| Greece | 40 | 20% | 8 |
| Brazil | 100 | 60% | 60 |
| Egypt | 220 | 40% | 88 |
| Argentina | 100 | 30% | 30 |

The total invested is, thus, 186.

This could have been obtained by the sum of four products:

In [88]:
+/ percent[2;] × invest[;3]÷100

186

In order to calculate the total invested for each partner and each year, we should repeat that algorithm for all the rows of `percent`, and all the columns of `invest`: this is precisely what an _inner product_ does.

And because it _adds_ a series of _products_, it will be expresseed by a dot (the operator) between a plus and a multiply sign, like this:

In [89]:
percent +.× invest÷100

90 113 104 125 176
164 182 186 249 396
116 155 170 226 348

In <!--figure-->the presentation below<!--Inner_Product-->, we have detailed the elementary products which lead to the calculation for bank 2 in year 3:

![Diagram representing the _inner product_ operation.](../res/Inner_Product.png)

<!--figure-->This figure<!--Inner_Product_Diagram--> has a great advantage: it clearly shows the relations that exist between the 3 matrices:

 - the left argument has as many columns as the right one has rows; and
 - the result has as many rows as the left argument and as many columns as the right one.

As you can see, the item `result[x;y]` is calculated from row `x` of the left argument (`⍺[x;]` in dfn notation) and column `y` of the right argument (`⍵[;y]` in dfn notation).

These rules will be generalised in the next section.

### Definition of Inner Product

The syntax of _inner product_ is `r ← x f.g y`, where the _inner product_ is represented by a dot (`.`) and `f` and `g` represent two appropriate dyadic functions (either primitive or user-defined).

The arguments may be arrays of any rank: scalars, vectors, matrices, or higher-rank arrays.
The shape of the arguments and the shape of the result follow very simple rules:

 - the length of the last dimension of the left argument must be equal to the length of the first dimension of the right argument (in other words, `(¯1↑⍴x)≡1↑⍴y`); and
 - the shape of the result is the catenation of the argugments' shapes, in which the common dimension has disappeared (in other words, `(⍴r)≡(¯1↓⍴x),1↓⍴y`).

Of course, as usual, scalars are repeated to fit the appropriate size.

Let us represent scalars by `s`, vectors by `v`, matrices by `m`, and higher-rank arrays by `a`.
The table below shows the same of the result of some _inner products_:

| `r ← x f.g y` | `⍴x` | `⍴y` | `⍴r` |
| :- | -: | -: | -: |
| `a ← a f.g a` | `2 3 8` | `8 5 4` | `2 3 5 4` |
| `a ← a f.g a` | `2 3 8` | `8 3 2` | `2 3 3 2` |
| `m ← m f.g m` | `3 5` | `5 8` | `3 8` |
| `v ← m f.g v` | `4 7` | `7` | `4` |
| `v ← v f.g m` | `4` | `4 7` | `7` |
| `s ← v f.g v` | `10` | `10` | `⍬` |

### Typical Uses of Inner Products

#### Two Simple Problems

Many students imagine that matrix products are complex things, reserved for mathematicians, and far removed from everyday life.
This opinion should be reconsidered: very simple problems can be solved using inner product.

`hms` is a variable which contains a time interval in hours, minutes, and seconds:

In [90]:
hms ← 3 44 29

We would like to convert it into seconds.
We shall see three methods just now, and a fourth method will be given in another chapter.

A horrible solution:

In [91]:
(3600×hms[1]) + (60×hms[2]) + hms[3]

13469

A good APL solution:

In [92]:
+/ 3600 60 1 × hms

13469

An excellent solution with inner product:

In [93]:
3600 60 1 +.× hms

13469

The second and third solutions are equivalent in terms of number of characters typed and similar in performance.
However, **it is recommended** that you use the third one: it will help you become familiar with _inner product_ so that after a certain period, it will become part of your toolkit as an APL programmer.

Here is a very similar example.
Two vectors represent the prices of a certain number of goods and the quantities we bought:

In [94]:
price ← 6 4.2 1.5 8.9 31 18
qty   ← 2 6   3   5   1  0.5

To calculate how much we paid, we can use the beginner's solution, or a solution with a simple _inner product_; they give the same result, of course:

In [95]:
+/ price × qty

126.2

In [96]:
price +.× qty

126.2

Just to show how it works, <!--figure-->the figure below<!--Inner_Product_Diagram_2--> contains an explanatory diagram similar to the one we used for our Banks/Investments example.

![Diagram explaining the behaviour of an _inner product_ between two vectors](res/Inner_Product_Diagram_2.png)

#### A Useful Family

Used with comparison functions, _inner product_ offers 18 extremely useful derived functions.

Here is a vector `ages` containing the ages of 400 persons:

In [97]:
⎕RL ← 73
⎕← 20↑ages ← ?400⍴100  ⍝ We display the first 20 ages only.

73 16 42 50 2 67 11 81 37 93 100 98 83 32 76 51 37 15 46 88

In the same way as we did in [a previous section](./Some-Primitive-Functions.ipynb#Reduction-of-Binary-Data), we can answer some elementary questions:

Are all these people younger than 65?

In [98]:
∧/ ages < 65

0

Is there at least one person younger than 20?

In [99]:
∨/ ages < 20

1

How many people are younger than 20?

In [100]:
+/ ages < 20

81

We can now replace _reduce_ in the previous examples by _inner product_, like this:

Are all these people younger than 65?

In [101]:
ages ∧.< 65

0

Is there at least one person younger than 20?

In [102]:
ages ∨.< 20

1

How many people are younger than 20?

In [103]:
ages +.< 20

81

Clever, isn't it?

These expressions can be read as:

 - `∧.<` means "all smaller" – are the ages **all smaller** than 65?
 - `∨.<` means "at least one is smaller" – is there **at least** one age **smaller** than 20?
 - `+.<` means "how many are smaller" – **how many** ages are **smaller** than 20?

In those three expressions, we have combined `∧`, `∨`, and `+` with `<`.
We could just as well combine them with all the comparison symbols, giving 18 different _inner products_, as shown in this table:

In [104]:
'∧∨+' ∘.{⍺,'.',⍵} '<≤=≥>≠'

┌───┬───┬───┬───┬───┬───┐
│∧.<│∧.≤│∧.=│∧.≥│∧.>│∧.≠│
├───┼───┼───┼───┼───┼───┤
│∨.<│∨.≤│∨.=│∨.≥│∨.>│∨.≠│
├───┼───┼───┼───┼───┼───┤
│+.<│+.≤│+.=│+.≥│+.>│+.≠│
└───┴───┴───┴───┴───┴───┘

#### A Special Case of a Comparison Inner Product

In this family of _inner products_, `∧.=` is particularly interesting, because it answers the question "are all those values equal?".
For example, applied to vectors of the same length:

In [105]:
'customer' ∧.= 'customer'

1

In [106]:
'customer' ∧.= 'cucumber'

0

Let us use this property to search for a word in a matrix of words:

In [107]:
⎕← words ← 8 7⍴'CONTACTCOLUMNSFORTUNEPRODUCTCOLONELPROVIDEMACHINETYPICAL'

CONTACT
COLUMNS
FORTUNE
PRODUCT
COLONEL
PROVIDE
MACHINE
TYPICAL

If we combine this 8 by 7 matrix with a 7-item vector, compatibility rules are obeyed, and the result will be a 8-item vector:

In [108]:
words ∧.= 'PRODUCT'

0 0 0 1 0 0 0 0

The shape of `words` is `8 7` and the shape of `'PRODUCT'` is `7`, so the common dimension disappears, and the result has shape `8`.

Now, let us search for three words:

In [109]:
⎕← three ← 3 7⍴'MACHINECOMFORTPRODUCT'

MACHINE
COMFORT
PRODUCT

In [110]:
words ∧.= ⍉three

0 0 0
0 0 0
0 0 0
0 0 1
0 0 0
0 0 0
1 0 0
0 0 0

We must transpose the matrix to be compliant with the compatibility rules, and the result shows what word was found in what row of `words`.

If we wanted to know which words were found, we could add an _or-reduction_:

In [111]:
∨⌿ words∧.=⍉three

1 0 1

If we wanted to know which of the rows in `words` contain words in `three`, we could have used another _or-reduction_, but along the other axis:

In [112]:
∨/ words∧.=⍉three

0 0 0 1 0 0 1 0

It may also be useful to search for the positions of said matches, but we can use _index of_ for that:

In [113]:
words ⍳ three

7 9 4

The converse to the expression `∧.=` is `∨.≠`.
(That means that `∧.=` and `∨.≠` always return opposite Boolean values.)
It looks for _different_ values instead of looking for _equal_ values.
Let us look at one simple example:

In [114]:
words ∨.≠ ⍉three

1 1 1
1 1 1
1 1 1
1 1 0
1 1 1
1 1 1
0 1 1
1 1 1

A `1` means that this word in `three` does not match the word in this row of `words`.
So, if a row contains all `1`s, the word in that row does not match any of the words in `three`.
Using _and-reduce_ along the second axis pinpoints the rows of `words` for which this is true:

In [115]:
∧/ words∨.≠⍉three

1 1 1 0 1 1 0 1

With a compression, we can see the words that are not found in `three`:

In [116]:
(∧/words∨.≠⍉three) ⌿ words

CONTACT
COLUMNS
FORTUNE
COLONEL
PROVIDE
TYPICAL

#### Similar Applications

Very often it is desirable to find out whether any rows (or columns) of a matrix contain all blanks or all zeroes; or, alternatively, whether any rows or columns contain at least one non-zero number or a non-blank character.

To solve the first task we can use the same _inner product_ as we used in most of the previous section (`∧.=`) and, to solve the second one, we can use the converse, which we introduced at the end of the previous section (`∨.≠`).

Suppose we have a matrix of characters `mc` and a matrix of numbers `mn`. Then,

 - `mc ∧.= ' '` says which rows contain all blanks;
 - `mn ∧.= 0` says which rows contain all zeroes;
 - `' ' ∧.= mc` says which columns contain all blanks;
 - `mc ∨.≠ ' '` says which rows contain at least one non-blank character;
 - `0 ∨.≠ mn` says which columns contain at least one non-zero number; and so on...

#### Shortest Routes in a Graph

Finding the shortest routes in a graph is a classical problem to which _inner product_ offers an elegant solution.
Imagine 6 points in a town.
They can be joined via a certain number of paths, according to <!--figure-->the figure below<!--Town_Diagram-->.

![A diagram with connections between 6 points in a town.](../res/Town_Diagram.png)

We can create a matrix with the distances between the points.
The missing paths will be represented by a very high value (`1000` in this case) to dissuade anyone from using them:

In [117]:
⍝ to            A    B    C    D    E    F 
distances ←,    0   21 1000   35 1000 1000  ⍝ from A to ...
distances ,←   32    0   34 1000 1000 1000  ⍝ from B to ...
distances ,← 1000   34    0 1000 1000   25  ⍝ from C to ...
distances ,← 1000   44 1000    0 1000 1000  ⍝ from D to ...
distances ,← 1000 1000   51   17    0 1000  ⍝ from E to ...
distances ,← 1000 1000   31 1000   24    0  ⍝ from F to ...
⎕← distances ← 6 6⍴distances

0 21 1000 35 1000 1000
 32 0 34 1000 1000 1000
1000 34 0 1000 1000 25
1000 44 1000 0 1000 1000
1000 1000 51 17 0 1000
1000 1000 31 1000 24 0

The matrix `distances` contains distances from one point to another one, assuming we follow a single path.
The values `1000` indicate connections that cannot be made directly.
For example, there is no direct route from `E` to `B`; can we get there in two steps?

Let us consider all the possible pairs of routes from `E` to `B`:

| Routes | Distances | Total distance |
| :- | -: | -: |
| `E → A` + `A → B` | 1000 + 21 | 1021 |
| `E → B` + `B → B` | 1000 + 0 | 1000 |
| `E → C` + `C → B` | 51 + 34 | 85 |
| `E → D` + `D → B` | 17 + 44 | 61 |
| `E → E` + `E → B` | 0 + 1000 | 1000 |
| `E → F` + `F → B` | 1000 + 1000 | 2000 |

These computations could have been performed by APL.
The first set of distances considered refer to routes that start at `E`, which are the values of the fifth row of the matrix `distances`:

In [118]:
distances[5;]

1000 1000 51 17 0 1000

Similarly, the second set of distances considered refer to routes that end at `B`, which are the values of the second column of the matrix `distances`:

In [119]:
distances[;2]

21 0 34 44 1000 1000

Therefore, all we have to do is sum the fifth row “from `E`” to the second column “to `B`” and we get distances of paths from `E`, with a possible intermediate stop, that end at `B`:

In [120]:
distances[5;] + distances[;2]

1021 1000 85 61 1000 2000

Only two routes really exist, because they are smaller than `1000`, which are the ones of length `85` and `61`.
Of course, we shall choose the shortest one:

In [121]:
⌊/ distances[5;]+distances[;2]

61

To obtain all the minimum routes that have one or two steps, we just have to repeat this calculation for all rows and columns:
an _inner product_ by the _minimum of sums_ will do that.

In [122]:
⎕← L2 ← distances ⌊.+ distances

0 21 55 35 1000 1000
 32 0 34 67 1000 59
 66 34 0 1000 49 25
 76 44 78 0 1000 1000
1000 61 51 17 0 76
1000 65 31 41 24 0

This result shows new routes, for example from `A` to `C`, from `B` to `F`, and others.

The matrix `distances` contains zeroes in the diagonal, because the distance from a point to itself is zero.
That means that the _inner product_ computed above shows a matrix that displays the length of the shortest route between two points that makes use of one or two roads.
That is why the matrix was called `L2`.

Notice that `L2` and `distances` have some common values:

In [123]:
L2 = distances

1 1 0 1 1 1
1 1 1 0 1 0
0 1 1 1 0 1
0 1 0 1 1 1
1 0 1 1 1 0
1 0 1 0 1 1

That is because the matrix `distances` already tells us what are the shortest routes between two points that make use of a **single** road.
Typically, going directly from a starting point to your final destination is faster than going through an intermediate stop, and that is why many values in `L2` match `distances`.

If we compute yet another _inner product_, we determine the lengths of the shortest routes between two points that make use of one, two, or three roads:

In [124]:
⎕← L3 ← L2 ⌊.+ distances

0 21 55 35 1000 80
32 0 34 67 83 59
66 34 0 66 49 25
76 44 78 0 1000 103
93 61 51 17 0 76
97 65 31 41 24 0

Again, `L3` and `L2` have some values in common:

In [125]:
L3 = L2

1 1 1 1 1 0
1 1 1 1 0 1
1 1 1 0 1 1
1 1 1 1 1 0
0 1 1 1 1 1
0 1 1 1 1 1

The values in common refer to connections for which considering a third road doesn't help.
For example, you couldn't go from `A → C` directly, but if you go through `B`, then you can go from `A → B → C`, and the length of that path is 55.
When computing `L3`, we try to see if there is a path `A → C`, using three roads, that is faster than the path we already know.
For example, is the path `A → D → B → C` faster?
In this case, it isn't, and that is why `L3[1;3]` and `L2[1;3]` are the same:

In [126]:
L3[1;3] = L2[1;3]

1

Imagine, however, that better roads are built in the directions `A → D` and `D → B`:

In [127]:
newDists ← distances
newDists[(1 4)(4 2)] ← 5 5
⎕← newDists

0 21 1000 5 1000 1000
 32 0 34 1000 1000 1000
1000 34 0 1000 1000 25
1000 5 1000 0 1000 1000
1000 1000 51 17 0 1000
1000 1000 31 1000 24 0

Now, using one or two roads, you can go from `A` to `C` in:

In [128]:
(newDists ⌊.+ newDists)[1;3]

55

However, if we consider routes that use one, two, or three roads, we can make use of the new better roads and go from `A` to `C` faster:

In [129]:
(newDists ⌊.+ newDists ⌊.+ newDists)[1;3]

44

If we ignore these new roads and continue using the distances we saw before, we can do one final _inner product_ between `L3` and `distances` to figure out how to connect all points:

In [130]:
⎕← L4 ← L3 ⌊.+ distances

0 21 55 35 104 80
32 0 34 67 83 59
66 34 0 66 49 25
76 44 78 0 127 103
93 61 51 17 0 76
97 65 31 41 24 0

`L4` does not contain the value `1000`, so we see it is possible to go from any point to any point using four roads or less.
If we tried doing an additional _inner product_, we would see that all values would stay the same, which means we already know the shortest routes possible:

In [131]:
L4 ≡ L4 ⌊.+ distances

1

Here is a representation of all the connections:

In [132]:
' ABCDEF','ABCDEF'⍪L4

A B C D E F
A 0 21 55 35 104 80
B 32 0 34 67 83 59
C 66 34 0 66 49 25
D 76 44 78 0 127 103
E 93 61 51 17 0 76
F 97 65 31 41 24 0

These computations using _inner product_ are elegant, but they have a shortcoming:
we found, for example, that the shortest path from `D` to `E` has length 127, and that it requires four steps, but we do not know which ones those four steps are.

#### Is a Graph Connected?

In some development projects involving large graphs, it is sometimes necessary to check whether all points can be reached from any other point.
In our town analogy from before, this would amount to checking whether we could go from any location to any other location, regardless of the length of the route.

We already know the answer for the diagram in <!--figure-->the figure above<!--Town_Diagram-->, so let us consider a modified version in <!--figure-->the figure below<!--Town_Diagram_Not_Connected-->:

![A modified diagram with connections between 6 points.](../res/Town_Diagram_Not_Connected.png)

To check if the graph is connected (to check if you can go from any point to any other point), we can represent connections with a `1` and absence of connections with a `0`:

In [133]:
connections ←, 1 1 0 1 0 0
connections ,← 1 1 0 0 0 0
connections ,← 0 0 1 0 0 1
connections ,← 0 1 0 1 0 0
connections ,← 0 0 1 1 1 0
connections ,← 0 0 1 0 1 1
connections ← 6 6⍴connections
' ABCDEF','ABCDEF'⍪connections

A B C D E F
A 1 1 0 1 0 0
B 1 1 0 0 0 0
C 0 0 1 0 0 1
D 0 1 0 1 0 0
E 0 0 1 1 1 0
F 0 0 1 0 1 1

In the matrix `distances`, the diagonal was set to `0` because the distance from a point to itself was `0`.
In the matrix `connections`, the diagonal is set to `1` because each point is (obviously!) connected to itself.

The matrix `connections` determines the direct connections between points, and we can see there is no direct connection from `C` to `E`:

In [134]:
connections[3;5]

0

Is there a connection in two steps?
This connection will exist if we can go from `C` to `A` **and** from `A` to `E`, **or** from `C` to `B` **and** from `B` to `E`, **or** ... and so on.

Repeated for all points, this connectivity matrix in two steps can be obtained using an _inner product_ by _or_ and _and_:

In [135]:
⎕← C2 ← connections ∨.∧ connections

1 1 0 1 0 0
1 1 0 1 0 0
0 0 1 0 1 1
1 1 0 1 0 0
0 1 1 1 1 1
0 0 1 1 1 1

Now, we can see that `C` is connected to `E`:

In [136]:
C2[3;5]

1

We can keep computing these _inner products_:

In [137]:
⎕← C3 ← C2 ∨.∧ connections

1 1 0 1 0 0
1 1 0 1 0 0
0 0 1 1 1 1
1 1 0 1 0 0
1 1 1 1 1 1
0 1 1 1 1 1

In [138]:
⎕← C4 ← C3 ∨.∧ connections

1 1 0 1 0 0
1 1 0 1 0 0
0 1 1 1 1 1
1 1 0 1 0 0
1 1 1 1 1 1
1 1 1 1 1 1

In [139]:
⎕← C5 ← C4 ∨.∧ connections

1 1 0 1 0 0
1 1 0 1 0 0
1 1 1 1 1 1
1 1 0 1 0 0
1 1 1 1 1 1
1 1 1 1 1 1

The matrix `C5` tells what points are connected by routes with 5 steps or less.
Do we need to compute another _inner product_?

Our town has 6 different points of interest.
If you cannot go from one place to another in 5 steps, there is no point in trying 6 steps:
routes with 6 steps will mean we are repeating intermediate stops, which does not help in trying to reach a specific destination.

So, our final connectivity matrix is `C5`:

In [140]:
C5

1 1 0 1 0 0
1 1 0 1 0 0
1 1 1 1 1 1
1 1 0 1 0 0
1 1 1 1 1 1
1 1 1 1 1 1

Because it contains `0`s, it shows that we cannot go from `D` to `C`, for example.
You can confirm this by looking at <!--figure-->the town diagram above<!--Town_Diagram_Not_Connected--> and realising there is, indeed, no roads that take us from `D` to `C`.

### Other Uses of Inner Product

We saw above some common uses of _inner product_, but there are many other useful _inner products_ using primitives or even user-defined functions.
Learning to recognise patterns where an _inner product_ is applicable is just a matter of practice.

As for _outer product_, some applications of _inner product_ produce nested arrays, as you can see with the following examples:

In [141]:
⎕← a ← 2 3⍴2 4 1 1 3 5

2 4 1
1 3 5

In [142]:
⎕← b ← 3 4⍴3 0 2 5 1 7 7 2 6 0 4 2

3 0 2 5
1 7 7 2
6 0 4 2

In [143]:
a ,.+ b

┌──────┬──────┬──────┬─────┐
│5 5 7 │2 11 1│4 11 5│7 6 3│
├──────┼──────┼──────┼─────┤
│4 4 11│1 10 5│3 10 9│6 5 7│
└──────┴──────┴──────┴─────┘

In [144]:
a +., b

┌────┬───┬────┬───┐
│7 10│7 7│7 13│7 9│
├────┼───┼────┼───┤
│9 10│9 7│9 13│9 9│
└────┴───┴────┴───┘

### Application

We have a certain number of two-dimensional points, the coordinates of which are given by a nested vector:

In [145]:
⍝ points:    A     B     C     D      E     F     G    H
coords ← (0 2)(¯1 2)(¯2 1)(¯1 0)(¯1 ¯1)(1 ¯3)(2 ¯2)(2 0)

Now, we take those coordinates and then split them into vectors with the x and y coordinates:

In [146]:
⎕← (x y) ← ↓[1]↑coords

┌───────────────────┬──────────────────┐
│0 ¯1 ¯2 ¯1 ¯1 1 2 2│2 2 1 0 ¯1 ¯3 ¯2 0│
└───────────────────┴──────────────────┘

<!--figure-->The figure below<!--Vertices_Polygon--> shows where those points are located in a coordinate system, and the polygon we get when we connect those points:

![A polygon with eight vertices.](../res/Vertices_Polygon.png)

The area of the polygon can be calculated by adding the areas of the trapeziums delimited by the polygon and the horizontal axis, like the grey trapezium $[FGHK]$.
For each of those trapeziums, their base lengths are calculated by subtracting adjacent values in `x`:

In [147]:
x - 1⌽x

1 1 ¯1 0 ¯2 ¯1 0 2

Then, the base lengths have to be multiplied by half of the sums of adjacent values in `y`:

In [148]:
(y + 1⌽y)÷2

2 1.5 0.5 ¯0.5 ¯2 ¯2.5 ¯1 1

In other words, we must _add_ the _products_ of bases by heights, which can be written as an _inner product_:

In [149]:
(x-1⌽x) +.× (y+1⌽y)÷2

11.5

What about the perimeter now?
We must add all the individual segment lengths.
The length of a segment $[BC]$ or $[FG]$ can be calculated using the Pythagorean theorem: $a^2 + b^2 = c^2$.

We shall calculate the length of horizontal and vertical sides by subtracting adjacent values in `x` and `y`, as we did for `x` in the previous example.

In [150]:
⎕← segs ← (x-1⌽x),[1.5](y-1⌽y)

1 0
 1 1
¯1 1
 0 1
¯2 2
¯1 ¯1
 0 ¯2
 2 ¯2

Now, in each small right-angled triangle, we must add the squares of both sides to obtain the square of each hypotenuse: _add_ the _squares_ will be our first _inner product_:

In [151]:
segs +.* 2

1 2 2 1 8 2 4 8

Then, we have to add the square roots of these hypotenuses.
_Add_ the _square roots_ will be our second _inner product_:

In [152]:
(segs +.* 2) +.* 0.5

13.8995

![](../res/Inner_Product_Drawing.png)

## Intermission Exercises

<!-- exercise reductions -->
***Exercise 2***:

You are given the matrix `mat` shown below.

Calculate the result of the following expressions and then check your answers on the computer:

 1. `⌈/ mat`
 2. `⌊/ +/ mat`
 3. `×/ ⌊/[1] mat`
 4. `×/ ⍴mat`
<!-- end -->

In [153]:
mat ← 3 5⍴8 2 5 1 4 3 7 1 5 0 4 3 6 0 6

<!-- exercise scans -->
***Exercise 3***:

Calculate:

 1. `-\ 6⍴1`
 2. `-\ ⌽⍳5`
 3. `×/ +\ 6⍴1`
<!-- end -->

<!-- exercise Boolean-scans-and-reductions -->
***Exercise 4***:

Calculate:

 1. `∧/ 1 1 1 0 1 1`
 2. `∧\ 1 1 1 0 1 1`
 3. `=/ 0 1 1 1 0 1 1`
 4. `=\ 0 1 1 1 0 1 1`
<!-- end -->

<!-- exercise reverse-engineer-scan -->
***Exercise 5***:

When we execute `×\vec`, we get the result `7 14 70 210 840`.
What is the value of `vec`?
<!-- end -->

<!-- exercise broken-keyboard-iota -->
***Exercise 6***:

Broken keyboard!
The iota (`⍳`) key of your keyboard does not work.
How could you create the list of the first `n` integers?
<!-- end -->

<!-- exercise triangle-area -->
***Exercise 7***:

Let $a$, $b$, and $c$, be the three sides of a triangle, and $s$ be its semi-perimeter (half of its perimeter, $s = \frac{a + b + c}2$).
Believe it or not, the area of that triangle is equal to the following expression (called “Heron's formula”):

$$\sqrt{s \times (s - a) \times (s - b) \times (s - c)}$$

Can you write a function to calculate the area of a triangle given the lengths of its sides?
Use the two examples below to validate your solution.
<!-- end -->

In [154]:
Area 3 4 5

VALUE ERROR: Undefined name: Area
      Area 3 4 5
      ∧


In [155]:
Area 10 6 8

VALUE ERROR: Undefined name: Area
      Area 10 6 8
      ∧


<!-- exercise all-different -->
***Exercise 8***:

We would like to know whether all the items of a vector are different.
Among the many possible solutions, could you find one using _outer product_ and another one using _inner product_?
The result must, of course, be a Boolean `0` or `1`.
<!-- end -->

<!-- exercise pairwise-reduction -->
***Exercise 9***:

What would be the result of `2 =/ 'MASSACHUSSETTS'`?
<!-- end -->

<!-- exercise word-in-vector -->
***Exercise 10***:

Try to find a word in a vector of characters.
Your function should give the positions of the first letter of the word in the vector.

After solving this exercise once, try to _also_ implement this function `In` using an _inner product_.
Solving the next two exercises first might help.
<!-- end -->

In [156]:
'CAN' In 'CAN YOU CANCEL MY FLIGHT ON AIR CANADA?'

VALUE ERROR: Undefined name: In
      'CAN'In'CAN YOU CANCEL MY FLIGHT ON AIR CANADA?'
           ∧


In [157]:
In ← {⍸∧⌿(¯1+⍳≢⍺)⌽⍺ ∘.= ⍵}

<!-- exercise find-row-in-matrix -->
***Exercise 11***:

Broken keyboard!
Try to find a row in a matrix without using dyadic iota `⍳`.
Your function `InMat` should expect a matrix on the left and a vector on the right, and it should return a Boolean value indicating whether the left argument matrix contains a row equal to the right argument vector.
You can assume the matrix has as many columns as the vector has elements.
<!-- end -->

In [158]:
InMat ← {∨/ ⍺ ∧.= ⍵}

In [159]:
(4 3⍴'BANCANDANDIG') InMat 'CAN'

1

In [160]:
(4 3⍴⍳12) InMat 0 1 2

0

<!-- exercise in-mat-find-better -->
***Exercise 12***:

Modify the previous exercise so that the right argument can either be a vector or a matrix.
Your function should return a vector with as many elements as there are rows on the right argument (or with one element, if the right argument is a vector), where each element is a Boolean value indicating whether the corresponding row of the right argument matches any of the rows of the left argument matrix.

Can you use _inner product_ in this exercise and/or in the previous one?
<!-- end -->

In [161]:
InMat2 ← {∨⌿⍺ ∧.= ⍉(¯2↑1,⍴⍵)⍴⍵}

In [162]:
(4 3⍴'BANCANDANDIG') InMat2 'CAN'

1

In [163]:
(3 3⍴9 8 7 4 5 6 3 2 1) InMat2 2 3⍴⍳6

0 1

<!-- exercise cross-count -->
***Exercise 13***:

For a certain number of people, you are given two vectors:

 1. `status` – a character vector specifying the person's marital status (Single, Married, Divorced, Widow, Unknown); and
 2. `graduated` – a character vector specifying if the person has graduated or not (Yes/No).

Write a function to count how many people are in each category.
<!-- end -->

In [164]:
CrossCount ← {binx ← (∪⍺)∘.=⍺ ⋄ biny ← ⍵∘.=(∪⍵) ⋄ r ← binx +.∧ biny ⋄ ((⊂''),(⊂'Total'),⍨∪⍵)⍪((⊂'Total'),⍨∪⍺),(⊢⍪+⌿)(⊢,+/)r}

In [165]:
⎕RL ← 73
⎕← 10↑status ← 'SMDWU'[?300⍴5]

SMMMDDUSUU

In [166]:
⎕RL ← 73
⎕← 10↑graduated ← 'YN'[?300⍴2]

YNNNNNYYYY

In [167]:
graduated CrossCount status

┌─────┬──┬──┬──┬──┬──┬─────┐
│ │S │M │D │U │W │Total│
├─────┼──┼──┼──┼──┼──┼─────┤
│Y │31│28│31│36│32│158 │
├─────┼──┼──┼──┼──┼──┼─────┤
│N │26│27│37│28│24│142 │
├─────┼──┼──┼──┼──┼──┼─────┤
│Total│57│55│68│64│56│300 │
└─────┴──┴──┴──┴──┴──┴─────┘

## At

The operator _at_ can be used to work with values at specified positions, and there are multiple ways in which we can do this.
In its most general form, the operator _at_ can work with right arguments of arbitrary rank, but for the sake of simplicity we will start by explaining how the operator works with vectors as right arguments.

### Apply Function At Indices

In one of its forms, the operator _at_ can be used to apply a function at the specified indices of the argument vector.
The function to be applied is the left operand and the indices are the right operand.

This is the first time we learn about an operator that takes an array as the right operand.
In the example below, notice how we use parentheses to separate the right operand array from the right argument array.

Let us _negate_ the values _at_ indices `1`, `3`, and `9`, of `vector`:

In [168]:
vector ← 10×⍳10
idx ← 1 3 9
(-@idx) vector

¯10 20 ¯30 40 50 60 70 80 ¯90 100

Notice that the operator _at_ returns the whole array after modifying the specified values, in opposition to returning just the values that were modified.

The parentheses around `-@idx` are necessary.
If they were not there, APL would not complain, but it would produce some interesting output:

In [169]:
-@idx vector

@
┌┴┐
- (1 3 9) (10+10×⎕io-⍨⍳10)

What is happening is that APL is looking at `idx vector` as a vector of length two defined by strand notation.
We can verify that `idx vector` is being interpreted as a single array.
We do that by parenthesising `idx vector` and noticing that the interesting output does not change:

In [170]:
-@(idx vector)

@
┌┴┐
- (1 3 9) (10+10×⎕io-⍨⍳10)

The exact meaning of this output will be explained in [the next chapter](./Tacit-Programming.ipynb#Tree-Diagram).
For now, it suffices to understand that it does _not_ the output of `(-@idx) vector`.

In order for APL to know that `idx` is the right _operand_ and `vector` is the right _argument_, we must separate them in some way.
Parenthesising the operator and its operand is one option, but another common option is to just use a _right tack_:

In [171]:
-@idx ⊢ vector

¯10 20 ¯30 40 50 60 70 80 ¯90 100

The _right tack_ works because it effectively breaks the strand, which lets the operator _at_ understand that its right operand is just `idx`.

Below, you can find another example usage of the operator _at_, in which we double values _at_ the indices `1` and `10` with a dfn:

In [172]:
{⍵×2}@1 10⊢ vector

20 20 30 40 50 60 70 80 90 200

The example above also shows that the operator _at_ works with user-defined functions.

It is important to understand how the operator _at_ passes the elements of the indices specified into the left operand function.
Take a look at the definition of the nested vector below:

In [173]:
⎕← nested ← ⍳¨⍳5

┌─┬───┬─────┬───────┬─────────┐
│1│1 2│1 2 3│1 2 3 4│1 2 3 4 5│
└─┴───┴─────┴───────┴─────────┘

What is the output you expect from running the expression `(⍴@4 5)nested`?
Give it some thought...
Because the actual result might surprise you:

In [174]:
⍴@4 5⊢nested

┌─┬───┬─────┬─┬─┐
│1│1 2│1 2 3│2│2│
└─┴───┴─────┴─┴─┘

The operator _at_ collects all the elements at the indices specified and then passes them, as a whole vector, to the left operand function.
We can see this by using, as the left operand, a dfn that prints its right argument:

In [175]:
_ ← {⎕←'called with ⍵: ' ⋄ ⎕←⍵ ⋄ ⍵}@4 5⊢nested

called with ⍵: 
┌───────┬─────────┐
│1 2 3 4│1 2 3 4 5│
└───────┴─────────┘

This is different from passing each element to the left operand function separately when the left operand function is **not** a _scalar function_, like `⍴`.

Then, after all the values are collected and passed in to the left operand function, the result should be a single scalar (that gets placed in all the indices specified) or a vector with as many elements as there were indices.
In that case, the elements of the result vector are distributed across the positions specified by the indices.

The example above with _shape at_ and the nested vector shows that a scalar result gets placed in all the positions that were specified.
The example below shows that the result vector must have as many elements as there were indices, otherwise the operator _at_ will not know how to place the results:

In [176]:
⎕← matrices ← {⍵ ⍵⍴⍳⍵*2}¨⍳3

┌─┬───┬─────┐
│1│1 2│1 2 3│
│ │3 4│4 5 6│
│ │ │7 8 9│
└─┴───┴─────┘

In [177]:
⍴@3⊢matrices

LENGTH ERROR
      ⍴@3⊢matrices
      ∧


The reason we get a `LENGTH ERROR` above is that the _shape at index `3`_ returns `3 3`, which is a vector of length `2`, which does not match the length `1` of the vector of indices.

### Apply Dyadic Function At Indices

When the function in `(F@idx)vector` is dyadic, the left argument of `F` can be written on the left of the operator _at_, as `left (F@idx) vector`.
In this case, the operator _at_ does not do any sort of selection or indexing into the left argument.
The left argument is passed directly to the function `F`:

In [178]:
10 100 ×@3 5 ⊢ vector

10 20 300 40 5000 60 70 80 90 100

Thus, if the function `F` is a scalar function, the left argument must be a scalar or have the same length as the vector of indices:

In [179]:
10 100 1000 ×@3 5 ⊢ vector

LENGTH ERROR: Mismatched left and right argument shapes
      10 100 1000×@3 5⊢vector
                 ∧


In [180]:
100 ×@1 3 5 ⊢ vector

1000 20 3000 40 5000 60 70 80 90 100

### Replace Values At Indices

Here is the nested vector from before:

In [181]:
nested

┌─┬───┬─────┬───────┬─────────┐
│1│1 2│1 2 3│1 2 3 4│1 2 3 4 5│
└─┴───┴─────┴───────┴─────────┘

Can you use a dfn to replace the elements of `nested` at indices `3` and `5` with the vectors `'CAT'` and `'DOG'`, respectively?

There are a couple of ways to do this.
For example, we can set the left operand to be a function that returns the left argument, and pass the new values as the left argument:

In [182]:
'CAT' 'DOG'{⍺}@3 5⊢nested

┌─┬───┬───┬───────┬───┐
│1│1 2│CAT│1 2 3 4│DOG│
└─┴───┴───┴───────┴───┘

This is equivalent to just enclosing the new values in a dfn in the left operand:

In [183]:
{'CAT' 'DOG'}@3 5⊢nested

┌─┬───┬───┬───────┬───┐
│1│1 2│CAT│1 2 3 4│DOG│
└─┴───┴───┴───────┴───┘

The operator _at_ can, indeed, be used to replace values at the specified indices.
We used a couple of different functions to do that, but the operator _at_ provides a special form, in which the left operand is an array instead of a function.

Suppose we have a vector `vec` and a vector with some indices, `idx`.
If the vector `new` has the same length as `idx`, then we have seen that `({new}@idx)vec` puts the elements of the vector `new` in the positions specified by the indices `idx`, but we can actually do this more directly by writing `(new@idx)vec`.
That is, the operator _at_ can take a vector on the left, instead of a function, and it will put those values in the positions specified:

In [184]:
'CAT' 'DOG'@3 5⊢nested

┌─┬───┬───┬───────┬───┐
│1│1 2│CAT│1 2 3 4│DOG│
└─┴───┴───┴───────┴───┘

The operator _at_ has another form that we discuss next.

### Right Operand as Selector

So far, the forms we have seen with the operator _at_ had an array right operand (more specifically, a vector).
The operator _at_ can also take a right operand selector function.
This function must return a Boolean mask that determines to which values the left operand is applied.

For example, the code below replaces all values larger than 5 with the value 0:

In [185]:
⎕← vector ← 10?10

10 7 6 5 4 8 2 3 1 9

In [186]:
IsLarger5 ← {⍵>5}

In [187]:
0@IsLarger5⊢ vector

0 0 0 5 4 0 2 3 1 0

When the right operand is a function, that function must be prepared to receive the whole right argument array at once.
In other words, the selector function is not applied individually to each element of the right argument.

So, if we define `LongerThan3`:

In [188]:
LongerThan3 ← {3<≢⍵}

And if we recall the vector `nested`:

In [189]:
nested

┌─┬───┬─────┬───────┬─────────┐
│1│1 2│1 2 3│1 2 3 4│1 2 3 4 5│
└─┴───┴─────┴───────┴─────────┘

We might have expected `({3↑⍵}@LongerThan3) nested` to trim the two last vectors.
Instead, it raises an error:

In [190]:
{3↑⍵}@LongerThan3⊢ nested

RANK ERROR: Right operand and argument ranks differ
      {3↑⍵}@LongerThan3⊢nested
           ∧


The error happens because `LongerThan3` takes the whole vector `nested` as its argument and produces a scalar result, the Boolean value `1`:

In [191]:
LongerThan3 nested

1

Then, the operator _at_ sees that result as the Boolean mask that will determine what values get passed in to the left operand function.
However, the Boolean mask is just a scalar `1` and the right argument is a vector, so there is a rank mismatch and the operator _at_ cannot do its magic.
If we modify the function `LongerThan3` to return its result as a vector, we would still have an error, but this time a `LENGTH ERROR`:

In [192]:
LongerThan3 ← {,3<≢⍵}
{3↑⍵}@LongerThan3⊢ nested

LENGTH ERROR: Right operand and argument lengths differ
      {3↑⍵}@LongerThan3⊢nested
           ∧


This time, the error is a `LENGTH ERROR` because the Boolean mask is the vector `,1` which has length `1`, whereas the argument vector `nested` has length `5`.

As we have seen, the operator _at_ is very versatile.
As a right operand, you can have:

 - the set of indices where you want to do some work; or
 - a selector function that produces a Boolean mask indicating the positions where you will do some work.

As a left operand, you can have:

 - a function to be applied to the subarray that was selected; or
 - new values to replace the subarray that was selected.

The left operand acts on the selection of the right operand.

### High-Rank Arguments

So far, all of our examples with the operator _at_ were with vectors.
However, the operator _at_ can handle right arguments of arbitrary rank.
We just need to understand how those high-rank arrays are handled by the operands of the operator _at_.

#### Right Operand Array

The right operand array can be a simple integer vector or a nested integer vector.
If it is a simple integer vector (or a simple integer scalar), then the integers are the indices of major cells of the argument.

For example, we can easily increment the second and fourth rows of a matrix:

In [193]:
⎕← mat ← 4 4⍴⍳16

1 2 3 4
 5 6 7 8
 9 10 11 12
13 14 15 16

In [194]:
1000 +@2 4⊢ mat

1 2 3 4
1005 1006 1007 1008
 9 10 11 12
1013 1014 1015 1016

If the right operand is nested, it specifies indices for _reach_ or _choose_ indexing.

For example, below we negate the four corners of our matrix `mat`:

In [195]:
-@((1 1)(1 4)(4 1)(4 4))⊢mat

¯1 2 3 ¯4
 5 6 7 8
 9 10 11 12
¯13 14 15 ¯16

#### Right Operand Function

When the right operand is a function, it must still compute a Boolean mask that determines what values will be modified or replaced.

For example, we can easily add one to the odd values in the matrix:

In [196]:
1 +@{2|⍵} mat

2 2 4 4
 6 6 8 8
10 10 12 12
14 14 16 16

#### Shape of the Left Operand

So far, all the examples that we saw and that involved high-rank arrays avoided the issue of shapes:

 - If the left operand is a function, what will be the shape of its argument?
 - If the left operand is an array, what should its shape be for _at_ to know how to replace the values?

Before you proceed and read this section, try to do some experiments to see if you can deduce these rules on your own.

Ready?

1. If the right operand is an array that specifies indices and the left operand is an array of replacement values, then the shape of the left operand must conform to the shape induced by the right operand.

If the right operand is an integer vector specifying major cells, the left operand must be of the same rank as the right argument, and it must have as many major cells as the length of the right operand:

In [197]:
(100+3 4⍴⍳4)@2 3 4⊢mat  ⍝ 3 indices, left operand has 3 rows.

1 2 3 4
101 102 103 104
101 102 103 104
101 102 103 104

If the right operand is a nested array that uses _reach_ or _choose indexing_ to select which values will be replaced, then the left operand must have the same shape as the array of indices:

In [198]:
corners ← (1 1)(1 4)(4 1)(4 4)

In [199]:
(⍳4)@corners⊢mat  ⍝ vector ←→ vector

1 2 3 2
5 6 7 8
9 10 11 12
3 14 15 4

In [200]:
(2 1 2⍴⍳4)@(2 1 2⍴corners)⊢mat  ⍝ 2 1 2 cuboid ←→ 2 1 2 cuboid

1 2 3 2
5 6 7 8
9 10 11 12
3 14 15 4

2. If the right operand is an array that specifies indices and the left operand is a function, then the right argument of the function will have the shape induced by the right operand (as seen in the previous case) and it must return an array of the same shape.

If we use _choose indexing_ with a two by two matrix to select the corners of our matrix `mat`, then the right argument of the left operand will be a two by two matrix and so must be the return value:

In [201]:
{⎕←'right arg:'⍵ ⋄ 2 2⍴⍳4}@(2 2⍴corners)⊢mat

┌──────────┬─────┐
│right arg:│ 1 4│
│ │13 16│
└──────────┴─────┘
1 2 3 2
5 6 7 8
9 10 11 12
3 14 15 4

Similarly, if we use a simple integer vector `vec` to select major cells of `array`, the right argument of the left operand will have shape equal to `(⍴vec),1↓⍴array` and the return value should have the same shape:

In [202]:
{⎕←'shape:'(⍴⍵) ⋄ 2 4⍴⍳4}@(2 3)⊢mat

┌──────┬───┐
│shape:│2 4│
└──────┴───┘
 1 2 3 4
 1 2 3 4
 1 2 3 4
13 14 15 16

3. If the right operand is a function that computes a Boolean mask, the values that are selected will be ravelled into a vector.
   1. If the left operand is a replacement array, such an array must be a vector with the same length as the ravel of the selected values.
   2. If the left operand is a function, its right argument will be the ravel of the values selected and it must return a vector with the same length.

The only exception to these two bullet points is if the left operand array is a scalar or if the left operand function returns a scalar.
In that case, all the values selected will be replaced by that scalar.

Let us demonstrate these rules by using the operator _at_ on the matrix `mat`:

In [203]:
mat

1 2 3 4
 5 6 7 8
 9 10 11 12
13 14 15 16

We will be working on the central two by two region of the matrix:

In [204]:
mat ∊ 6 7 10 11

0 0 0 0
0 1 1 0
0 1 1 0
0 0 0 0

We can replace those four values with four other values, if those are in a vector:

In [205]:
(100×⍳4)@{⍵∊6 7 10 11}mat

1 2 3 4
 5 100 200 8
 9 300 400 12
13 14 15 16

We can transform those four values, as long as the left operand returns a vector of length four:

In [206]:
{1⌽⍵}@{⍵∊6 7 10 11}mat  ⍝ transform those 4 values

1 2 3 4
 5 7 10 8
 9 11 6 12
13 14 15 16

The left operand can also be, or return, a scalar.
In that case, the scalar will replace all the selected values:

In [207]:
0@{⍵∊6 7 10 11}mat  ⍝ replace all with 0

1 2 3 4
 5 0 0 8
 9 0 0 12
13 14 15 16

## Rank Operator

The operator _rank_ is represented by the character jot diaeresis `⍤`, and you can type it with the keys <kbd>APL</kbd>+<kbd>Shift</kbd>+<kbd>J</kbd>.
A good mnemonic is that it is the jot (<kbd>APL</kbd>+<kbd>j</kbd>) with something on top (<kbd>Shift</kbd>).

The operator _rank_ is a powerful operator that allows you to apply functions to sub-arrays of your argument arrays.
The function to be applied is the left operand `F` and the sub-arrays to which `F` is applied are defined by the right operand `spec`.

### Sub-Arrays in the Monadic Case

Typically, the best way of helping new learners grasp how the operator _rank_ works is by exploring the result of `⊂⍤spec` for various integer scalars `spec` and for the case where `⊂⍤spec` is used monadically.

With that in mind, we define a cuboid below:

In [208]:
⎕← cuboid ← 2 3 4⍴⍳24

1 2 3 4
 5 6 7 8
 9 10 11 12

13 14 15 16
17 18 19 20
21 22 23 24

The operator _rank_ in `⊂⍤spec ⊢ cuboid` will apply the function _enclose_ to the cells of `cuboid` that have rank specified by `spec`:

 - if `spec` is `0`, the operator _rank_ will enclose all the scalars in the cuboid because scalars have rank `0`;
 - if `spec` is `1`, the operator _rank_ will enclose all the rows of the cuboid because rows have rank `1`;
 - if `spec` is `2`, the operator _rank_ will enclose all the planes / sub-matrices in the cuboid because planes have rank `2`; and
 - if `spec` is `3` or more, the operator _rank_ will enclose the whole cuboid because the cuboid has rank `3`.

In [209]:
⊂⍤3⊢cuboid

┌───────────┐
│ 1 2 3 4│
│ 5 6 7 8│
│ 9 10 11 12│
│ │
│13 14 15 16│
│17 18 19 20│
│21 22 23 24│
└───────────┘

In [210]:
⊂⍤2⊢cuboid

┌──────────┬───────────┐
│1 2 3 4│13 14 15 16│
│5 6 7 8│17 18 19 20│
│9 10 11 12│21 22 23 24│
└──────────┴───────────┘

In [211]:
⊂⍤1⊢cuboid

┌───────────┬───────────┬───────────┐
│1 2 3 4 │5 6 7 8 │9 10 11 12 │
├───────────┼───────────┼───────────┤
│13 14 15 16│17 18 19 20│21 22 23 24│
└───────────┴───────────┴───────────┘

In [212]:
⊂⍤0⊢cuboid

1 2 3 4
 5 6 7 8
 9 10 11 12

13 14 15 16
17 18 19 20
21 22 23 24

As these examples show, the operator _rank_ led the function _enclose_ (its left operand) to the sub-arrays that had the rank that was chosen by the right operand.

Suppose that we wanted to transpose each of the two sub-matrices of the cuboid.
One can achieve that effect with dyadic transpose, but we can also look at that as _transpose rank 2_, which means we transpose the cells that have rank 2:

In [213]:
⍉⍤2⊢cuboid

1 5 9
 2 6 10
 3 7 11
 4 8 12

13 17 21
14 18 22
15 19 23
16 20 24

### Sub-Arrays in the Dyadic Case

The derived function of the operator _rank_ can also be used dyadically.
In that case, the operator _rank_ will pair up sub-arrays of the left argument with sub-arrays of the right argument.
The sub-arrays of both sides do **not** have to be of the same rank.

In general, the dyadic case looks like `left (F⍤i j) right`.
The integer scalar `i` determines the rank of the sub-arrays of the left argument `left` and the integer scalar `j` determines the rank of the sub-arrays of the right argument `right`.

In order for the operator _rank_ to be able to pair up sub-arrays of the left argument with sub-arrays of the right argument, there has to be some shape compatibility:
the _frames_ of the left and right argument must be the same.
The _frames_ are composed of the leading axes of the arguments that are not included in the sub-arrays:

 - the _frame_ of the left argument is `fl ← (-i)↓⍴left`; and
 - the _frame_ of the right argument is `fr ← (-j)↓⍴right`.

So, `left F⍤i j ⊢ right` only works if `fl ≡ fr`, that is, if `((-i)↓⍴left)≡((-j)↓⍴right)`.

For example, suppose `left` has rank `5` and shape `2 3 4 5 6` and `right` has rank `4` and shape `2 4 3 8`:

In [284]:
left ← 2 3 4 5 6⍴0
right ← 2 4 3 8⍴0

 - the expression `left F⍤4 3 ⊢ right` is a valid usage of the operator _rank_ because:
    - the _frame_ of `left` is `2`, which is `¯4↓⍴left`; and
    - the _frame_ of `right` is also `2`, which is `¯3↓⍴right`.

In [286]:
_← left {⍺ ⍵}⍤4 3 ⊢ right  ⍝ works without error

 - the expression `left F⍤4 2 ⊢ right` does not work and gives a `RANK ERROR` because the _frames_ of the left and right arguments do not have the same rank:
    - the _frame_ of `left` is `2`, which is `¯4↓⍴left`; and
    - the _frame_ of `right` is `2 4`, which is `¯3↓⍴right`.

In [288]:
_← left {⍺ ⍵}⍤4 2 ⊢ right  ⍝ RANK ERROR

RANK ERROR
      _←left{⍺ ⍵}⍤4 2⊢right  ⍝ RANK ERROR
                 ∧


 - the expression `left F⍤3 2 ⊢ right` does not work and gives a `LENGTH ERROR` because the _frames_ of the left and right arguments do not have the same axes lengths:
    - the _frame_ of `left` is `2 3`, which is `¯3↓⍴left`; and
    - the _frame_ of `right` is `2 4`, which is `¯2↓⍴right`.

In [296]:
_← left {⍺ ⍵}⍤3 2 ⊢ right  ⍝ LENGTH ERROR

LENGTH ERROR: It must be that either the left and right frames match or one of them has length 0
      _←left{⍺ ⍵}⍤3 2⊢right  ⍝ LENGTH ERROR
                 ∧


The error message above shows one more situation in which the operator _rank_ works.
If the sub-arrays of the left argument match the whole left argument or if the sub-arrays of the right argument match the whole right argument, then the operator _rank_ will pair that argument with all the sub-arrays of the other argument.

 - the expression `left F⍤5 j ⊢ right` works for any value of `j` because the frame of `left` is `⍬` and therefore has length zero.

In [302]:
_← left {⍺ ⍵}⍤5 0 ⊢ right  ⍝ works!
_← left {⍺ ⍵}⍤5 1 ⊢ right  ⍝ works!!
_← left {⍺ ⍵}⍤5 2 ⊢ right  ⍝ works!!!
_← left {⍺ ⍵}⍤5 3 ⊢ right  ⍝ works!!!!

Similarly,

 - the expression `left F⍤i 4 ⊢ right` works for any value of `i` because the frame of `right` is `⍬` and therefore has length zero.

In [303]:
_← left {⍺ ⍵}⍤0 4 ⊢ right  ⍝ works!
_← left {⍺ ⍵}⍤1 4 ⊢ right  ⍝ works!!
_← left {⍺ ⍵}⍤2 4 ⊢ right  ⍝ works!!!
_← left {⍺ ⍵}⍤3 4 ⊢ right  ⍝ works!!!!
_← left {⍺ ⍵}⍤4 4 ⊢ right  ⍝ works!!!!!

To illustrate better how the pairing of the operator _rank_ works, we can consider the dfn `{⍺ ⍵}` that will strand the left and right sub-arrays.

First, we will pair all sub-matrices on the left with all scalars on the right.

In [298]:
(3 4 5⍴⎕A) {⍺ ⍵}⍤2 0 ⊢ ⍳3

┌─────┬─┐
│ABCDE│1│
│FGHIJ│ │
│KLMNO│ │
│PQRST│ │
├─────┼─┤
│UVWXY│2│
│ZABCD│ │
│EFGHI│ │
│JKLMN│ │
├─────┼─┤
│OPQRS│3│
│TUVWX│ │
│YZABC│ │
│DEFGH│ │
└─────┴─┘

Now, we will pair all rows on the left with the right argument vector.

In [301]:
(3 4 5⍴⎕A) {⍺ ⍵}⍤1 1 ⊢ ⍳3

┌─────┬─────┐
│ABCDE│1 2 3│
├─────┼─────┤
│FGHIJ│1 2 3│
├─────┼─────┤
│KLMNO│1 2 3│
├─────┼─────┤
│PQRST│1 2 3│
└─────┴─────┘
┌─────┬─────┐
│UVWXY│1 2 3│
├─────┼─────┤
│ZABCD│1 2 3│
├─────┼─────┤
│EFGHI│1 2 3│
├─────┼─────┤
│JKLMN│1 2 3│
└─────┴─────┘
┌─────┬─────┐
│OPQRS│1 2 3│
├─────┼─────┤
│TUVWX│1 2 3│
├─────┼─────┤
│YZABC│1 2 3│
├─────┼─────┤
│DEFGH│1 2 3│
└─────┴─────┘

Notice how, in the example above, the right argument vector was repeated 12 times, one for each of the sub-arrays of the left argument.

### Cells

The sub-arrays of rank `r` of an array are typically referred to as `r`-cells of that array.

 - `0`-cells are scalars;
 - `1`-cells are vectors and sometimes referred to as rows;
 - `2`-cells are matrices and sometimes referred to as planes or tables;
 - ...

So, the expression `F⍤k ⊢ array` can be said to apply `F` to the `k`-cells of `array`.
We can say the same thing the other way around: the `k`-cells of `array` are its sub-arrays on the last `k` axes.

#### Negative `k`

If `k` is negative, a `k`-cell of an array of rank `r` is the same as a `r+k`-cell.
Using negative values of `k` can be advantageous because this makes it so that the actual rank of the cell depends on the rank of the original array.
For example, a `¯2`-cell on a matrix is actually a `(2+¯2)`-cell (a scalar), and a `¯2`-cell on a cuboid is actually a `(3+¯1)`-cell (a vector).

#### Major Cells

`¯1`-cells are commonly referred to as _major cells_ because they are the largest cells of an array, as far as rank is considered.
Below, you can find a table with common arrays and the corresponding major cells.

| array | major cells |
| :- | :- |
| vector | scalars |
| matrix | vectors |
| cuboid | matrices |
| 4D array | cuboids |

### Specifying Cells

Now that we are aware of what _cells_ are, the operator _rank_ can actually be defined in terms of _cells_ of the left and right arguments.

The expression `F⍤k ⊢ right` will apply the function `F` monadically to the `k`-cells of `right` and the expression `left F⍤i j ⊢ right` will apply the function `F` dyadically between the `i`-cells of `left` and the `j`-cells of `right`.
In particular, the operator _rank_ can deal with _cell_ specifications that have negative numbers.

For example, the expression `⊂⍤¯1` turns any array into a vector of its _major cells_.

If we have a matrix:

In [306]:
3 3⍴⍳9

1 2 3
4 5 6
7 8 9

we get a vector of its rows:

In [307]:
⊂⍤¯1 ⊢ 3 3⍴⍳9

┌─────┬─────┬─────┐
│1 2 3│4 5 6│7 8 9│
└─────┴─────┴─────┘

And if we have a cuboid:

In [308]:
2 3 4⍴⍳24

1 2 3 4
 5 6 7 8
 9 10 11 12

13 14 15 16
17 18 19 20
21 22 23 24

we get a vector of its sub-matrices:

In [309]:
⊂⍤¯1 ⊢ 2 3 4⍴⍳24

┌──────────┬───────────┐
│1 2 3 4│13 14 15 16│
│5 6 7 8│17 18 19 20│
│9 10 11 12│21 22 23 24│
└──────────┴───────────┘

The operator _rank_ has another form that is slightly more general, but we will only cover it in [the Specialist's Section](#The-Full-Rank-Specification).

## Key

## Stencil

## More Exercises

<!-- exercise trim-long-subvectors -->
Write a function that receives a vector of vectors and returns the same vector of vectors, but all the subvectors that had a length of four or more were replaced by their first three elements.
How would you have to implement this function if you wanted to use the operator _at_?
<!-- end -->

In [219]:
TrimLong ← {({3↑¨⍵}@({3<≢⍵}¨))⍵}

In [220]:
TrimLong ⍳¨⍳5

┌─┬───┬─────┬─────┬─────┐
│1│1 2│1 2 3│1 2 3│1 2 3│
└─┴───┴─────┴─────┴─────┘

## The Specialist's Section

<br />
<center><i>Each chapter is followed by a "Specialist's Section" like this one. This section is dedicated to skilled APLers who wish to improve their knowledge.</i>

You will find here rare or complex usages of the concepts presented in this chapter, or discover extended explanations which need the knowledge of some symbols that will be seen much further in the book.

<b>If you are exploring APL for the first time, skip this section and go to the next chapter.</b></center>

### Rank Outer Products

### The Full Rank Specification

## Solutions

The following solutions we propose are not necessarily the “best” ones; perhaps you will find other solutions that we have never considered. APL is a very rich language, and due to the general nature of its primitive functions and operators there are always plenty of different ways to express different solutions to a given problem. Which one is “the best” depends on many things, for example the level of experience of the programmer, the importance of system performance, the required behaviour in border cases, the requirement to meet certain programming standards and also personal preferences. This is one of the reasons why APL is so pleasant to teach and to learn!

We advise you to try and solve the exercises before reading the solutions!

<!-- solution ex-complex-refunding -->
***Solution 1***:

We start by defining a numeric vector with the upper limits and a matrix with the refund rates:

In [221]:
limits ← 600 1100 1500 2000
⎕← 4 0⍕ rates ← 3 4⍴100 100 80 50 100 70 30 10 80 60 20 5

100 100 80 50
 100 70 30 10
 80 60 20 5

The next step is rewriting the traditional refund rate table in the modified version.
For example, for category 1 students, expenses up to 2.000 are refunded in 50%, and then there is a 30% additional refund for expenses up to 1.500.
Why is that?
Because the refund rate for expenses up to 1.500 is 80%, and not 50%, meaning there is a difference of 30%.

By using an _n-wise reduction_, we can subtract each pair of adjacent columns and get these rates:

In [222]:
⎕← modifiedRates ← 2 -/ rates,0

0 20 30 50
30 40 20 10
20 40 15 5

In the expression above, we added a column of zeroes to the right of `rates`.
This can be interpreted as "all expenses above 2.000 have a refund rate of 0%".

Now, we compare each student's expenses with the upper limits of each range:

In [223]:
capped ← expenses ∘.⌊ limits

In [224]:
10↑expenses,' ',capped

2121 600 1100 1500 2000
1936 600 1100 1500 1936
 682 600 682 682 682
 818 600 818 818 818
1922 600 1100 1500 1922
1419 600 1100 1419 1419
1389 600 1100 1389 1389
 465 465 465 465 465
1829 600 1100 1500 1829
 221 221 221 221 221

Finally, we are ready to multiply each student's capped expenses with the appropriate rates.
We just have to be careful to take into account each student's category to use the appropriate rates:

In [225]:
refunds ← +/capped × modifiedRates[categories;] ÷ 100
⎕← 10↑refunds
⎕← +/refunds

1670 1113.6 657.4 610.8 1112.2 843.8 1036.7 372 1584.5 221

309491

To wrap up, we can put everything into a function:

In [226]:
∇ refunds ← Refund (limits rates categories expenses)
    ;modifiedRates ;capped
    modifiedRates ← 2 -/ rates,0
    capped ← expenses ∘.⌊ limits
    refunds ← +/capped × modifiedRates[categories;] ÷ 100
∇

In [227]:
10↑Refund limits rates categories expenses

1670 1113.6 657.4 610.8 1112.2 843.8 1036.7 372 1584.5 221

In [228]:
+/Refund limits rates categories expenses

309491

<!-- solution reductions -->
***Solution 2***:

In [229]:
⌈/ mat

4 8 12 16

In [230]:
⌊/ +/ mat

10

In [231]:
×/ ⌊/[1] mat

24

In [232]:
×/ ⍴mat

16

<!-- solution scans -->
***Solution 3***:

In [233]:
-\ 6⍴1

1 0 1 0 1 0

In [234]:
-\ ⌽⍳5

5 1 4 2 3

In [235]:
×/ +\ 6⍴1

720

<!-- solution Boolean-scans-and-reductions -->
***Solution 4***:

In [236]:
∧/ 1 1 1 0 1 1

0

In [237]:
∧\ 1 1 1 0 1 1

1 1 1 0 0 0

In [238]:
=/ 0 1 1 1 0 1 1

1

In [239]:
=\ 0 1 1 1 0 1 1

0 0 0 0 1 1 1

<!-- solution reverse-engineer-scan -->
***Solution 5***:

The result of `×\vec` has six elements, which means the original vector `vec` also has six elements.
The first element of a _scan_ is the first element of the argument vector, thus the first element of `vec` is `7`.
Finally, each subsequent element of the result is going to be the previous element of the result _times_ the respective element of the original vector `vec`.

For example, `14` is `7 × vec[2]`, meaning `vec[2] ← 14 ÷ 7`.
Then, `70` should be `14 × vec[3]`, meaning `vec[3] ← 70 ÷ 14`.
And we can keep going, or we can compute everything with a pairwise reduction:

In [240]:
⎕← vec ← 7, ¯2 ÷/ 7 14 70 210 840

7 2 5 3 4

In [241]:
×\ vec  ⍝ verification

7 14 70 210 840

<!-- solution broken-keyboard-iota -->
***Solution 6***:

To create a vector with the first `n` integers we can create a vector with `n` elements `1` and then do a _plus scan_ on that:

In [242]:
Iota ← {+\⍵⍴1}

In [243]:
Iota 10

1 2 3 4 5 6 7 8 9 10

<!-- solution triangle-area -->
***Solution 7***:

We can translate the formula directly to APL and add an _inner product_ for good measure:

In [244]:
Area ← {s ← 0.5×+/⍵ ⋄ (s×s×.-⍵)*0.5}

In [245]:
Area 3 4 5

6

In [246]:
Area 10 6 8

24

<!-- solution all-different -->
***Solution 8***:

The _outer product_ is commonly used when you need to try all different combinations of something, so if you have a vector `v`, you can use `v ∘.= v` to compare all items against each other:

In [247]:
v ← ⍳4

In [248]:
v ∘.= v

1 0 0 0
0 1 0 0
0 0 1 0
0 0 0 1

If a vector only contains unique vectors, then the resulting Boolean matrix will look like the one shown above: a matrix filled with zeroes with a main diagonal composed of ones.

In this context, it suffices to guarantee that each row, or each column, contains a single `1`, which we can do as follows:

In [249]:
1∧.=+/v∘.=v

1

This way, we found a solution that uses both an _inner product_ and an _outer product_.

<!-- solution pairwise-reduction -->
***Solution 9***:

The _pairwise equals reduction_ will verify which adjacent characters are the same:

In [250]:
2 =/ 'MASSACHUSSETTS'

0 0 1 0 0 0 0 0 1 0 0 1 0

<!-- solution word-in-vector -->
***Solution 10***:

The first natural step is comparing each character of the left argument to the whole right argument vector:

In [251]:
word ← 'CAN'
sentence ← 'CAN YOU CANCEL MY FLIGHT ON AIR CANADA?'
sentence⍪ word ∘.= sentence

C A N Y O U C A N C E L M Y F L I G H T O N A I R C A N A D A ?
1 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0
0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 1 0 1 0
0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0

Then, we see an interesting pattern: the left argument word occurs in the positions where the matrix has a diagonal of ones.
So, what we need to do is look for those diagonals of ones in the matrix that is the result of the _outer product_.
This is not very easy to do directly, but we can _rotate_ the rows of the matrix independently to try to align the diagonals:

In [252]:
(¯1+⍳≢word)⌽word ∘.= sentence

1 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0
1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 1 0 1 0 0
1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0

Then we can use an _and reduction_ to verify if the diagonals were complete or not:

In [253]:
⍸∧⌿(¯1+⍳≢word)⌽word ∘.= sentence

1 9 33

Putting everything together, we can define our dfn `In`:

In [254]:
In ← {⍸∧⌿(¯1+⍳≢⍺)⌽⍺∘.=⍵}

In [255]:
'CAN' In 'CAN YOU CANCEL MY FLIGHT ON AIR CANADA?'

1 9 33

To solve this exercise with an _inner product_, we can do something different, but similar. But different.

First, we create a matrix with as many rows as the length of the left argument vector, where the sentence in each row has been rotated one character in comparison to the row above:

In [256]:
l ← ≢word
(¯1+⍳l)⌽l(≢sentence)⍴sentence

CAN YOU CANCEL MY FLIGHT ON AIR CANADA?
AN YOU CANCEL MY FLIGHT ON AIR CANADA?C
N YOU CANCEL MY FLIGHT ON AIR CANADA?CA

Next, we use the _inner product_ `∧.=` to look for columns of that matrix that match the word:

In [257]:
⍸word∧.=(¯1+⍳l)⌽l(≢sentence)⍴sentence

1 9 33

<!-- solution find-row-in-matrix -->
***Solution 11***:

One possible way of tackling this problem is by using the _axis_ operator together with _equals_ to compare each row with the vector given:

In [258]:
(4 3⍴'BANCANDANDIG') =[2] 'CAN'

0 1 1
1 1 1
0 1 1
0 0 0

Then, we can use an _and reduction_ to find rows that are filled with `1`:

In [259]:
∧/ (4 3⍴'BANCANDANDIG') =[2] 'CAN'

0 1 0 0

Finally, we use an _or reduction_ to determine if _any_ of the rows matched:

In [260]:
∨/∧/(4 3⍴'BANCANDANDIG') =[2] 'CAN'

1

In [261]:
InMat ← {∨/∧/⍺=[2]⍵}

In [262]:
(4 3⍴⍳12) InMat 0 1 2

0

Another alternative revolves around understanding that the pattern of _and reduction_ after _equals_ can be written as an _inner product_ that will pair the vector with each row of the matrix:

In [263]:
(4 3⍴'BANCANDANDIG')∧.='CAN'

0 1 0 0

After that, we only need the final _or reduction_ to produce the final Boolean result:

In [264]:
InMat ← {∨/⍺∧.=⍵}

In [265]:
(4 3⍴'BANCANDANDIG') InMat 'CAN'

1

In [266]:
(4 3⍴⍳12) InMat 0 1 2

0

<!-- solution in-mat-find-better -->
***Solution 12***:

We can follow the same _inner product_ strategy as in the previous exercise, we just have to make sure we match the dimensions of the left and right arguments of the _inner product_ correctly.
First, notice that we do _have_ to do something:

In [267]:
(3 3⍴9 8 7 4 5 6 3 2 1) InMat 2 3⍴⍳6

LENGTH ERROR
InMat[0] InMat←{∨/⍺∧.=⍵}
                    ∧


When we learned about the _inner product_, we learned that the last dimension of the left argument must match the first dimension of the right argument.
So, in order for this _inner product_ to work, we actually need to transpose the right argument:

In [268]:
(3 3⍴9 8 7 4 5 6 3 2 1) ∧.= ⍉2 3⍴⍳6

0 0
0 1
0 0

By transposing the right argument, we fix this issue.
In a way, when the left and right arguments to an _inner product_ are matrices, it is as if we were combining the rows of the left matrix with the columns of the right matrix.
In the case of the _inner product_ `∧.=`, it is as if we were doing an equality comparison between the rows of the left matrix and the columns of the right matrix.

After transposing the right argument, we need to use _reduce-first_ to determine which rows of the right argument appear in the left argument:

In [269]:
∨⌿ (3 3⍴9 8 7 4 5 6 3 2 1) ∧.= ⍉2 3⍴⍳6

0 1

What is also worth noting is that this new implementation will also work when the right argument is just a vector, because the _transpose_ will do nothing in that case, and the final reduction is the same, regardless of it being a _first-axis_ or _last-axis_ reduction:

In [270]:
∨⌿ (4 3⍴'BANCANDANDIG') ∧.=⍉ 'CAN'

1

Thus, our solution can be:

In [271]:
InMat2 ← {∨⌿⍺∧.=⍉⍵}

<!-- solution cross-count -->
***Solution 13***:

The first step to counting the people in each category is creating two Boolean matrices that encode the category to which each person belongs.
We start by determining which categories exist and then we use an _outer product_ to create said matrices:

In [272]:
∪status

SMDUW

In [273]:
10↑ maritalMat ← status ∘.= ∪status

1 0 0 0 0
0 1 0 0 0
0 1 0 0 0
0 1 0 0 0
0 0 1 0 0
0 0 1 0 0
0 0 0 1 0
1 0 0 0 0
0 0 0 1 0
0 0 0 1 0

In [274]:
10↑ graduateMat ← graduated ∘.= ∪graduated

1 0
0 1
0 1
0 1
0 1
0 1
1 0
1 0
1 0
1 0

Now, we can use an _inner product_ to _add up_ how many people have a specific marital status _and_ a specific graduation status:

In [275]:
(⍉graduateMat)+.∧maritalMat

31 28 31 36 32
26 27 37 28 24

We can write this in a dfn and then add the totals row and column, as well as the labels:

In [276]:
]dinput
CrossCount ← {
    binw ← ∪⍵
    bina ← ∪⍺
    mat ← (bina∘.=⍺)+.∧⍵∘.=binw
    mat ,← +/ mat
    mat ⍪← +⌿ mat
    (' ',binw,⊂'Total')⍪(bina,⊂'Total'),mat
}

In [277]:
graduated CrossCount status

┌─────┬──┬──┬──┬──┬──┬─────┐
│ │S │M │D │U │W │Total│
├─────┼──┼──┼──┼──┼──┼─────┤
│Y │31│28│31│36│32│158 │
├─────┼──┼──┼──┼──┼──┼─────┤
│N │26│27│37│28│24│142 │
├─────┼──┼──┼──┼──┼──┼─────┤
│Total│57│55│68│64│56│300 │
└─────┴──┴──┴──┴──┴──┴─────┘